{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ed372500-fb27-472e-a92d-469d22cfc2ef",
   "metadata": {},
   "source": [
    "## Detection of ORF regions from homology search\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "eab565f7-4cff-423b-b075-0e81027b9708",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "import pandas as pd\n",
    "import math\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "\n",
    "\n",
    "command = \"mafft-linsi ../03_Extended_homolog_search/HRR25_merged.fna > ./HRR25_mafft_linsi.txt\"\n",
    "#os.system(command)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f067ae4f-2814-4dcc-8cec-8005180ee7d5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "20e5f9c2-d6c4-4f48-b66b-486e280636a9",
   "metadata": {},
   "source": [
    "# Evaluating gaps in the alignment file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "dc0fc9f0-829b-4c92-9660-f8717b068c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def fasta2dict(f):\n",
    "    d = {}\n",
    "    with  open(f,\"r\") as F:\n",
    "        for line in F:\n",
    "            if line[0]==\">\":\n",
    "                name = line[1:].split(\"\\n\")[0].split(\"|\")[0]\n",
    "                d[name] = ''\n",
    "            else:\n",
    "                #print(name)\n",
    "                d[name]+= line.split(\"\\n\")[0]\n",
    "    F.close()\n",
    "    return d\n",
    "\n",
    "\n",
    "def MSAdict2df_AA(d):\n",
    "    \"\"\"\n",
    "    Dictionary of Multiple Sequence Alignment to pandas dataframe for evaluating gap regions.\n",
    "    \"\"\"\n",
    "    n_entries = len(d)\n",
    "    n_chars   = max([len(d[k]) for k in d])\n",
    "\n",
    "    df = pd.DataFrame(data={k:list(d[k]) for k in d} )\n",
    "    \n",
    "    try:\n",
    "        gap = df.apply(pd.Series.value_counts, axis=1).fillna(0)[\"-\"]    \n",
    "    except KeyError:\n",
    "        gap = pd.Series(data=0, index=[i for i in range(n_chars)])\n",
    "        \n",
    "    \n",
    "    df.insert(0,'Position', [i+1 for i in range(n_chars)])\n",
    "    df.insert(1,'GapRatio', [(float(i)/float(n_entries)) for i in gap])\n",
    "    df2 = df.swapaxes(\"index\", \"columns\")\n",
    "    return df2 \n",
    "\n",
    "\n",
    "\n",
    "def MSAdict2df_nt(d):\n",
    "    \"\"\"\n",
    "    Dictionary of Multiple Sequence Alignment to pandas dataframe for evaluating gap regions.\n",
    "    \"\"\"\n",
    "    n_entries = len(d)-1\n",
    "    n_chars   = max([len(d[k]) for k in d])\n",
    "\n",
    "    df = pd.DataFrame(data={k:list(d[k]) for k in d} )\n",
    "     \n",
    "    try:\n",
    "        gap = df.apply(pd.Series.value_counts, axis=1).fillna(0)[\"-\"]    \n",
    "    except KeyError:\n",
    "        gap = pd.Series(data=0, index=[i for i in range(n_chars)])\n",
    "    \n",
    "    df.insert(0,'GapRatio', [(float(i+0)/float(n_entries)) for i in gap])\n",
    "    \n",
    "    df2 = df.swapaxes(\"index\", \"columns\")\n",
    "    #df2[\"Homolog\"] = df2.index\n",
    "    df2.reset_index(inplace = True)\n",
    "    return df2 \n",
    "\n",
    "\n",
    "def df2fasta(df,f):\n",
    "    with open(f,\"w\") as F:\n",
    "        for index, row in df.iterrows():\n",
    "            gene = index\n",
    "            seq  = (\"\").join([row[i] for i in df.columns.values.tolist()])\n",
    "            entry = \">%s\\n%s\\n\"%(gene,seq)\n",
    "            F.write(entry)\n",
    "        F.close()\n",
    "    print(\"Generated fasta: %s\"%(f))\n",
    "        \n",
    "    \n",
    "def df2phy(df,f):\n",
    "    nrow  = len(df.index)\n",
    "    nchar = len(df. columns)\n",
    "    df = df.drop([\"Position\"])\n",
    "    \n",
    "    with open(f,\"w\") as F:\n",
    "        F.write(\"%d   %d\\n\\n\"%(nrow,nchar))\n",
    "        \n",
    "        for index, row in df.iterrows():\n",
    "            gene = index\n",
    "            seq  = (\"\").join([row[i] for i in df.columns.values.tolist()])\n",
    "            entry = \"%s      %s\\n\"%(gene,seq)\n",
    "            F.write(entry)\n",
    "        F.close() \n",
    "        \n",
    "    \n",
    "\n",
    "\n",
    "def rev_comp(seq):\n",
    "    complement_dict = {'A':'T','T':'A','G':'C','C':'G','N':'N'}\n",
    "    return \"\".join([complement_dict[base.upper()] for base in reversed(seq)])\n",
    "\n",
    "def translate_dna2aa(dna_seq):\n",
    "    dna_seq = dna_seq.upper()\n",
    "    codon_table = {\n",
    "    'GCT':\"A\",\n",
    "    'GCC':\"A\",\n",
    "    'GCA':\"A\",\n",
    "    'GCG':\"A\",\n",
    "    'TGT':\"C\",\n",
    "    'TGC':\"C\",\n",
    "    'GAT':\"D\",\n",
    "    'GAC':\"D\",\n",
    "    'GAA':\"E\",\n",
    "    'GAG':\"E\",\n",
    "    'TTT':'F',\n",
    "    'TTC':'F',\n",
    "    'GGT':'G',\n",
    "    'GGC':'G',\n",
    "    'GGA':'G',\n",
    "    'GGG':'G',\n",
    "    'CAT':'H',\n",
    "    'CAC':'H',\n",
    "    'ATT':'I',\n",
    "    'ATC':'I',\n",
    "    'ATA':'I',\n",
    "    'AAA':'K',\n",
    "    'AAG':'K',\n",
    "    'TTA':'L',\n",
    "    'TTG':'L',\n",
    "    'CTT':'L',\n",
    "    'CTC':'L',\n",
    "    'CTA':'L',\n",
    "    'CTG':'L',\n",
    "    'ATG':'M',\n",
    "    'AAT':'N', 'AAC':'N',\n",
    "    'CCT':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P',\n",
    "    'CAA':'Q', 'CAG':'Q',\n",
    "    'CGT':'R', 'CGC':'R', 'CGA':'R', 'CGG':'R', 'AGA':'R', 'AGG':'R',\n",
    "    'TCT':'S', 'TCC':'S', 'TCA':'S', 'TCG':'S', 'AGT':'S', 'AGC':'S',\n",
    "    'ACT':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T',\n",
    "    'GTT':'V', 'GTC':'V', 'GTA':'V', 'GTG':'V',\n",
    "    'TGG':'W',\n",
    "    'TAT':'Y', 'TAC':'Y',\n",
    "    'TAG':\"*\",'TGA':\"*\",'TAA':\"*\",\n",
    "\n",
    "    #Inverse table for the standard genetic code (compressed using IUPAC notation)\n",
    "    'GCN':\"A\",\n",
    "    'CGN':\"R\",'AGR':\"R\",'CGY':\"R\",'MGR':\"R\",\n",
    "    'AAY':\"N\",\n",
    "    'GAY':\"D\",\n",
    "    'RAY':\"B\",\n",
    "    'TGY':\"C\",\n",
    "    'CAR':\"Q\",\n",
    "    'GAR':\"E\",\n",
    "    'SAR':\"Z\",\n",
    "    'GGN':\"G\",\n",
    "    'CAY':\"H\",\n",
    "    'ATH':\"I\",\n",
    "    'CTN':\"L\",'TTR':\"L\",'CTY':\"L\",'YTR':\"L\",\n",
    "    'AAR':\"K\",\n",
    "    'TTY':\"F\",\n",
    "    'CCN':\"P\",\n",
    "    'TCN':\"S\",'AGY':\"S\",\n",
    "    'ACN':\"T\",\n",
    "    'TAY':\"Y\",\n",
    "    'GTN':\"V\",\n",
    "    'TRA':\"*\",'TAR':\"*\"\n",
    "                }\n",
    "\n",
    "    aa_seq = \"\"\n",
    "    for i in range(1,int(len(dna_seq)/3)+1):\n",
    "        try:\n",
    "            aa_seq += codon_table[dna_seq[(i-1)*3:i*3]]\n",
    "        except KeyError:\n",
    "            #print(\"Undefined codon: '%s' was translated to X \"%(dna_seq[(i-1)*3:i*3]))\n",
    "            aa_seq += 'X'\n",
    "    return(aa_seq)\n",
    "\n",
    "def fasta2dict(f):\n",
    "    d = {}\n",
    "    with  open(f,\"r\") as F:\n",
    "        for line in F:\n",
    "            if line[0]==\">\":\n",
    "                name = line[1:].split(\"\\n\")[0].split(\"|\")[0]\n",
    "                d[name] = ''\n",
    "            else:\n",
    "                #print(name)\n",
    "                d[name]+= line.split(\"\\n\")[0]\n",
    "    F.close()\n",
    "\n",
    "    return d\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "def make_fasta_from_listoftups(L,name):\n",
    "    with open(\"%s\"%name,\"w\") as f:\n",
    "        for I in L:\n",
    "            #print I\n",
    "            f.write('>%s\\n'%(str(I[0])))\n",
    "            f.write('%s\\n'%(str(I[1])))\n",
    "        f.close()\n",
    "        #print(\"Made fna file : %s\"%(name))\n",
    "\n",
    "\n",
    "def get_blastout(x):\n",
    "    files = []\n",
    "    lis = os.listdir(\"%s\" %x)\n",
    "    for i in lis:\n",
    "        if i[-4:] == \".out\" :\n",
    "            files.append(i)\n",
    "    return list(files)\n",
    "\n",
    "def parse_blastout(dir_out):\n",
    "    lis = os.listdir(dir_out)\n",
    "    fs = [\"%s/%s\" %(dir_out,f) for f in lis]\n",
    "    D = {}\n",
    "    for f in fs:\n",
    "        if f.split(\"/\")[-1][0] != \".\":\n",
    "            genome = f.split(\"/\")[-1].split(\".blast\")[0]\n",
    "            with open(f, \"r\") as F:\n",
    "                hit_n = 0\n",
    "                for line in F:\n",
    "                    hit_n+=1\n",
    "                    hit         = hit_n   \n",
    "                    cols        =    line.split(\",\")\n",
    "                    contig      =    cols[0]\n",
    "                    subject     =    cols[1]\n",
    "                    bit         =    float(cols[2])\n",
    "                    str_on_q    =    int(cols[6])\n",
    "                    end_on_q    =    int(cols[7])\n",
    "                    str_on_s    =    int(cols[8])\n",
    "                    end_on_s    =    int(cols[9])\n",
    "                    evalue      =    float(cols[10])\n",
    "\n",
    "\n",
    "                    el = {}\n",
    "                    el[\"contig\"]   = contig\n",
    "                    el[\"subject\"]    = subject\n",
    "                    el[\"str_s\"]      = str_on_s\n",
    "                    el[\"end_s\"]      = end_on_s\n",
    "                    el[\"str_q\"]      = str_on_q\n",
    "                    el[\"end_q\"]      = end_on_q\n",
    "                    el[\"evalue\"]   = evalue\n",
    "                    el[\"bitscore\"] = bit\n",
    "                    try:\n",
    "                        D[genome][hit] = el\n",
    "                    except KeyError:        \n",
    "                        D[genome] = {}\n",
    "                        D[genome][hit] = el\n",
    "\n",
    "            F.close()\n",
    "    return D    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "e4b21aae-a690-4c6a-a3b1-953aefb64cf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reference genes used as query for homology search\n",
    "query_homologs = fasta2dict(\"../03_Extended_homolog_search/blast/db/HRR25_nuc_nonAligned.fasta\")\n",
    "query          = list(query_homologs.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "14986cf2-3b00-4f2c-81d0-62f04c798452",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>10513</th>\n",
       "      <th>10514</th>\n",
       "      <th>10515</th>\n",
       "      <th>10516</th>\n",
       "      <th>10517</th>\n",
       "      <th>10518</th>\n",
       "      <th>10519</th>\n",
       "      <th>10520</th>\n",
       "      <th>10521</th>\n",
       "      <th>10522</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GapRatio</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.714286</td>\n",
       "      <td>0.807143</td>\n",
       "      <td>0.835714</td>\n",
       "      <td>0.835714</td>\n",
       "      <td>0.835714</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.992857</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.985714</td>\n",
       "      <td>0.978571</td>\n",
       "      <td>0.971429</td>\n",
       "      <td>0.971429</td>\n",
       "      <td>0.964286</td>\n",
       "      <td>0.964286</td>\n",
       "      <td>0.964286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_030569995.1_ASM3056999v1_genomic.fna_Conti...</td>\n",
       "      <td>a</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>c</td>\n",
       "      <td>a</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>c</td>\n",
       "      <td>a</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tetrapisispora_phaffii.fas_Contig:NC_016524.1_...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>t</td>\n",
       "      <td>a</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>a</td>\n",
       "      <td>t</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tetrapisispora_phaffii.fas_Contig:NC_016529.1_...</td>\n",
       "      <td>a</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>g</td>\n",
       "      <td>t</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_900074765.1_LAFE0_genomic.fna_Contig:LT598...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>a</td>\n",
       "      <td>c</td>\n",
       "      <td>c</td>\n",
       "      <td>c</td>\n",
       "      <td>t</td>\n",
       "      <td>t</td>\n",
       "      <td>g</td>\n",
       "      <td>g</td>\n",
       "      <td>t</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10524 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               index         0         1  \\\n",
       "0                                           GapRatio  0.714286  0.714286   \n",
       "1  GCA_030569995.1_ASM3056999v1_genomic.fna_Conti...         a         t   \n",
       "2  tetrapisispora_phaffii.fas_Contig:NC_016524.1_...         -         -   \n",
       "3  tetrapisispora_phaffii.fas_Contig:NC_016529.1_...         a         t   \n",
       "4  GCA_900074765.1_LAFE0_genomic.fna_Contig:LT598...         -         -   \n",
       "\n",
       "          2         3         4         5         6         7         8  ...  \\\n",
       "0  0.714286  0.714286  0.714286  0.807143  0.835714  0.835714  0.835714  ...   \n",
       "1         g         c         a         a         c         c         a  ...   \n",
       "2         -         -         -         -         -         -         -  ...   \n",
       "3         g         g         t         t         g         t         g  ...   \n",
       "4         -         -         -         -         -         -         -  ...   \n",
       "\n",
       "  10513     10514 10515     10516     10517     10518     10519     10520  \\\n",
       "0   1.0  0.992857   1.0  0.985714  0.978571  0.971429  0.971429  0.964286   \n",
       "1     -         -     -         -         -         -         -         -   \n",
       "2     -         -     -         t         a         t         g         a   \n",
       "3     -         -     -         -         -         -         -         -   \n",
       "4     -         a     c         c         c         t         t         g   \n",
       "\n",
       "      10521     10522  \n",
       "0  0.964286  0.964286  \n",
       "1         -         -  \n",
       "2         t         t  \n",
       "3         -         -  \n",
       "4         g         t  \n",
       "\n",
       "[5 rows x 10524 columns]"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mafft_out = \"HRR25_mafft_linsi.txt\"\n",
    "msa_d     = fasta2dict(mafft_out)\n",
    "msa_df = MSAdict2df_nt(msa_d)\n",
    "msa_df.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "bafcdfca-2208-4d34-a5a0-63896905483f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>10513</th>\n",
       "      <th>10514</th>\n",
       "      <th>10515</th>\n",
       "      <th>10516</th>\n",
       "      <th>10517</th>\n",
       "      <th>10518</th>\n",
       "      <th>10519</th>\n",
       "      <th>10520</th>\n",
       "      <th>10521</th>\n",
       "      <th>10522</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>Saccharomyces_pastorianus_GRS66_005211</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>Saccharomyces_pastorianus_GRS66_010945</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Saccharomyces_cerevisiae_YPL204W</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Saccharomyces_eubayanus_DI49_5263</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Saccharomyces_bayanus_var.uvarum_Suva_16.102</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 10524 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           index  0  1  2  3  4  5  6  7  8  \\\n",
       "93        Saccharomyces_pastorianus_GRS66_005211  -  -  -  -  -  -  -  -  -   \n",
       "94        Saccharomyces_pastorianus_GRS66_010945  -  -  -  -  -  -  -  -  -   \n",
       "95              Saccharomyces_cerevisiae_YPL204W  -  -  -  -  -  -  -  -  -   \n",
       "96             Saccharomyces_eubayanus_DI49_5263  -  -  -  -  -  -  -  -  -   \n",
       "97  Saccharomyces_bayanus_var.uvarum_Suva_16.102  -  -  -  -  -  -  -  -  -   \n",
       "\n",
       "    ... 10513 10514 10515 10516 10517 10518 10519 10520 10521 10522  \n",
       "93  ...     -     -     -     -     -     -     -     -     -     -  \n",
       "94  ...     -     -     -     -     -     -     -     -     -     -  \n",
       "95  ...     -     -     -     -     -     -     -     -     -     -  \n",
       "96  ...     -     -     -     -     -     -     -     -     -     -  \n",
       "97  ...     -     -     -     -     -     -     -     -     -     -  \n",
       "\n",
       "[5 rows x 10524 columns]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_query = msa_df[msa_df['index'].isin(query)]\n",
    "df_query.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "e104d69f-2f15-4b68-a4b6-41bacd7ec544",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "49"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_query)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "de7accc3-ae6e-4796-91cc-99fbde12b0b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-</th>\n",
       "      <th>Eremothecium_cymbalariae_Ecym_2202</th>\n",
       "      <th>Eremothecium_gossypii_AGOS_AFL091W</th>\n",
       "      <th>Kazachstania_africana_KAFR_0C02290</th>\n",
       "      <th>Kazachstania_africana_KAFR_0K00320</th>\n",
       "      <th>Kazachstania_barnettii_KABA2_04S02882</th>\n",
       "      <th>Kazachstania_barnettii_KABA2_06S03410</th>\n",
       "      <th>Kazachstania_exigua_HRR25_1</th>\n",
       "      <th>Kazachstania_exigua_HRR25_2</th>\n",
       "      <th>Kazachstania_naganishii_KNAG_0D01130</th>\n",
       "      <th>...</th>\n",
       "      <th>Zygosaccharomyces_bailii_FUGC01000003.1_cds_SJM83735.1_1</th>\n",
       "      <th>Zygosaccharomyces_parabailii_ZPAR0C01310_A</th>\n",
       "      <th>Zygosaccharomyces_parabailii_ZPAR0K01220_B</th>\n",
       "      <th>Zygosaccharomyces_rouxii_ZYRO0G08866g</th>\n",
       "      <th>Zygotorulaspora_mrakii_HG535_0B00830</th>\n",
       "      <th>a</th>\n",
       "      <th>c</th>\n",
       "      <th>g</th>\n",
       "      <th>t</th>\n",
       "      <th>Pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>49.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>49.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>49.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>49.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>49.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      -  Eremothecium_cymbalariae_Ecym_2202  \\\n",
       "0  49.0                                 0.0   \n",
       "1  49.0                                 0.0   \n",
       "2  49.0                                 0.0   \n",
       "3  49.0                                 0.0   \n",
       "4  49.0                                 0.0   \n",
       "\n",
       "   Eremothecium_gossypii_AGOS_AFL091W  Kazachstania_africana_KAFR_0C02290  \\\n",
       "0                                 0.0                                 0.0   \n",
       "1                                 0.0                                 0.0   \n",
       "2                                 0.0                                 0.0   \n",
       "3                                 0.0                                 0.0   \n",
       "4                                 0.0                                 0.0   \n",
       "\n",
       "   Kazachstania_africana_KAFR_0K00320  Kazachstania_barnettii_KABA2_04S02882  \\\n",
       "0                                 0.0                                    0.0   \n",
       "1                                 0.0                                    0.0   \n",
       "2                                 0.0                                    0.0   \n",
       "3                                 0.0                                    0.0   \n",
       "4                                 0.0                                    0.0   \n",
       "\n",
       "   Kazachstania_barnettii_KABA2_06S03410  Kazachstania_exigua_HRR25_1  \\\n",
       "0                                    0.0                          0.0   \n",
       "1                                    0.0                          0.0   \n",
       "2                                    0.0                          0.0   \n",
       "3                                    0.0                          0.0   \n",
       "4                                    0.0                          0.0   \n",
       "\n",
       "   Kazachstania_exigua_HRR25_2  Kazachstania_naganishii_KNAG_0D01130  ...  \\\n",
       "0                          0.0                                   0.0  ...   \n",
       "1                          0.0                                   0.0  ...   \n",
       "2                          0.0                                   0.0  ...   \n",
       "3                          0.0                                   0.0  ...   \n",
       "4                          0.0                                   0.0  ...   \n",
       "\n",
       "   Zygosaccharomyces_bailii_FUGC01000003.1_cds_SJM83735.1_1  \\\n",
       "0                                                0.0          \n",
       "1                                                0.0          \n",
       "2                                                0.0          \n",
       "3                                                0.0          \n",
       "4                                                0.0          \n",
       "\n",
       "   Zygosaccharomyces_parabailii_ZPAR0C01310_A  \\\n",
       "0                                         0.0   \n",
       "1                                         0.0   \n",
       "2                                         0.0   \n",
       "3                                         0.0   \n",
       "4                                         0.0   \n",
       "\n",
       "   Zygosaccharomyces_parabailii_ZPAR0K01220_B  \\\n",
       "0                                         0.0   \n",
       "1                                         0.0   \n",
       "2                                         0.0   \n",
       "3                                         0.0   \n",
       "4                                         0.0   \n",
       "\n",
       "   Zygosaccharomyces_rouxii_ZYRO0G08866g  \\\n",
       "0                                    0.0   \n",
       "1                                    0.0   \n",
       "2                                    0.0   \n",
       "3                                    0.0   \n",
       "4                                    0.0   \n",
       "\n",
       "   Zygotorulaspora_mrakii_HG535_0B00830    a    c    g    t  Pos  \n",
       "0                                   0.0  0.0  0.0  0.0  0.0    0  \n",
       "1                                   0.0  0.0  0.0  0.0  0.0    1  \n",
       "2                                   0.0  0.0  0.0  0.0  0.0    2  \n",
       "3                                   0.0  0.0  0.0  0.0  0.0    3  \n",
       "4                                   0.0  0.0  0.0  0.0  0.0    4  \n",
       "\n",
       "[5 rows x 55 columns]"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occurance_count = df_query.apply(pd.Series.value_counts, axis='rows').fillna(0)\n",
    "\n",
    "occurance_count2 = occurance_count.swapaxes(\"index\", \"columns\")\n",
    "\n",
    "#count_gaps = occurance_count.loc['-']\n",
    "#count_gaps.head()\n",
    "occurance_count2[\"Pos\"] = occurance_count2.index\n",
    "occurance_count2= occurance_count2.drop(index=[\"index\"])\n",
    "occurance_count2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "1f34f294-ea2a-4cd3-aba2-f961fd64d028",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:xlabel='Pos'>"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXAAAAEGCAYAAAB8Ys7jAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/MnkTPAAAACXBIWXMAAAsTAAALEwEAmpwYAAAeE0lEQVR4nO3de5Qc5Xnn8e8zPaMZ3YWk0QVGYoQQd7CEBwzGwYAQ91jYaxwc2xEbvFrHxMHYexywz4kPjhOTZJfN2uusTYiJ4js2dlBwOEYRxvZ6MTACgQQSFldphJAGCQnQDUnz7B9dM9PTl+nLVHVX1fw+53Cmu7q6+nkH+PU7b731lrk7IiKSPE2NLkBERGqjABcRSSgFuIhIQinARUQSSgEuIpJQzfX8sOnTp3tnZ2c9P1JEJPHWrFnzmru352+va4B3dnbS3d1dz48UEUk8M3u52HYNoYiIJFRFPXAzewl4EzgCHHb3LjObCvwQ6AReAj7k7q9HU6aIiOSrpgd+obsvdPeu4PnNwGp3XwCsDp6LiEidjGQIZSmwIni8Arh6xNWIiEjFKg1wBx4wszVmtjzYNtPdtwEEP2cUe6OZLTezbjPr7u3tHXnFIiICVD4L5Tx3f8XMZgCrzGxjpR/g7ncAdwB0dXVp5SwRkZBU1AN391eCnzuAnwJnA9vNbDZA8HNHVEWKiEihsj1wMxsPNLn7m8HjS4AvASuBZcBtwc97oypy9YbtPLlld1SHF4mNnXvf5oSZE1n27s6K9t978DD3PN7DH5w1h9bmTNWfd/djW7jlp+v44/M6+cKVpwx57UfdW3hiy24ef/l1PnvJicxvH89x7RMGXr9/3TbmTB3HsdPG8eM1PSw+aSa/2tRLa3MTW17fzycvmM8//7+XaMk08Z/f3UlTkwGwdfd+Hn1xJ6/uOcjmXfton9hKW0sTH3xnB+u37qH3zYPMmNjGhSfNGPics+dNZdqEVnrfPMhdv3mRh1/YyQu9e9mz/xB/fN48JrRmeLJnDwvnTKGaJbLfOniEB555lQ8sOqbsvm1jMhx4+wgPv7CT9y08ht43Dgy8NnncGABWPvkKJ86cwNotu7n01FmsemY7l5wyE4D3n9nBvOnjK66tElausWZ2HNleN2QD/3vu/ldmNg24G5gLbAaucfddwx2rq6vLa7mQ5y/uXc+3f1t0HrtIqvT/7/i7L1/OmObyfyB/85fP85X7N3L7h97BB87sqPrzOm/+2cDjp2+9lPGtzUVf6/fSbVcOeX3y2Bb+5IL53HZ/4ajq9z7+Lv7wzkcA+I/PnM/xMyYC8JE7f8tvnttZsP/xMybw3I63hnzWrr1vc+ZfrmLR3Cn89JPncfsDz/LVB58r2y6zsrsAg7/vcu8pFZNmpV/L3++u687ighOLniqs4P22JmcG4ICyPXB3fwF4R5HtO4HFNVVTpS8tPY0vLT2tHh8l0lD9oelU1ot8NegFvr7v0Ig/u6+Gm7vs2X+I1948WPS1/YeODDw+3Dd47EdeKN7P27xzX8G2w0f6ANiyaz8ABw73FX3vjz5xLtd842EAPn3xAj598QkVVA8f+ubDPPriLr7/X87h3PnTSu73xXvXs+LhoZ3IpQuP5n9du4ifP/0q//Xbawres2DGBDbteIvPX3ESy8+fX1E91dKVmCICUOFXRlZfX/m9384J20q+G4p+gQz0irOvtWSKd5PHZAajrJK/XAbryh63qUyPvf+Y48YMDlP1f2Y1nxe2uq6FIiLV27JrH1t372fzrn1ccfpsJrQW/9/2vqde4eTZk5ifM05djfvXbePCk2awcu0rTB0/pug+d/76Bd44cHhI4H3nkeLDm3/3wLMDj/vDef3WPUN647nyt39t9SauPXsuAK+99TYv79zLk1v2FH1vbojmhnk5/d8ZVmbMpf/441ub2ff2kSHbWkt8Xn9rorxrpQJcJOZ+729/MfB46+v7uWlJ4fDA4SN9/On3nuD4GRP4j8+8t6bP+fN71jGmuWlIzznfl3+2oWDbgUPF93+hd+/A4/4Q+8zdayuu53+s+t1AgAPc9MO1PL55d8F+Ny5ewKxJbQPPj2uv/EThded10v3y6xw/Y/gvvf7Xz+6cys/WbRuyreOocUPGwhfNncITm3fz2SUn8Ll7nuI9C6ZXXE+1FOAiCbJ5V+E4MUB/5zX3JGAthgvvkegPt1f3HBh+xzy5HeP1W98oeH3u1HEDX2jP//UV9LnTUkUP/KozjuaqM44uu9/7F3Vw5elH05IxvuqLONLnAz3wudPGsenLlwPQnPfZl502q2zvfiQU4CIpUOlJz0bpr6+CofMhcqPvSJGxiNx2Z5qMDNGFZX9gZyz7Wbnyg7tflOENOokpInXwJ995nIee3cFbBw9X9b6Nr7458PhIkfSPcnw5CRTgIilgEfY8w7B1936uu+uxqt/3kWAeeSkKcBGRQO40OYk/BbhICoQ1Bt4/RU6SQQEuIpJQCnARkYRSgIuIJJQCXCSG4j6rROJBAS4iklAKcBEZYmyLphImhQJcJAXCvKBlfGuGd8yZEt4Ba1RqRUQZpAAXkSHMjMljWxpdhv4SqIACXESGaDJicQq1ucTNG2SQAlxEhmgyK3uHmnpojkMRMacAF5EhjOiXQa1EUwxqiDsFuIgMYTHpgSvAy1OAi8gQTU0x6YHH4Vsk5hTgIjJEk8XjOlDld3kKcBEZInsSs/HpGYca4k4BLpICHuKVPGZDbyYc5nGroR54eQpwkRhqZOczqh54tUeMwzh83CnARWKo2g51mGGXnUYY2uFqph54eQpwkQQJc6iklCazSHq/1R5TY+DlKcBFUiCMYO+/8tEsmt5vtYfUNMLyFOAiCRLluHB/YGoaYXIowEUEGAzMpqZohi+qPeTeg0dCryFtKg5wM8uY2RNmdl/wfKqZrTKzTcHPo6IrU0SGE8bQeH9oRzUGXi2tB15eNT3wG4ENOc9vBla7+wJgdfBcRBIqY/1j4BbNPPAqB2YyGkMpq6IAN7MO4ErgzpzNS4EVweMVwNWhViYiFQsjcPuPYUQ0/lzlMRXf5VXaA/974HNAX862me6+DSD4OSPc0kSknjI5s1DicBozBqM4sVc2wM3sKmCHu6+p5QPMbLmZdZtZd29vby2HEJEywh4Db4pgekP1eawEL6eSf03nAe8zs5eAHwAXmdl3gO1mNhsg+Lmj2Jvd/Q5373L3rvb29pDKFpGw5c67juZCnmj3H43KBri73+LuHe7eCVwLPOjuHwVWAsuC3ZYB90ZWpYgMK4zrMzM5iRnNELgSOWwj+UPpNmCJmW0ClgTPRSQEjeh9Ng05idn4sG18BfHXXM3O7v4Q8FDweCewOPySRKQR+odQnHgMX8ShhrjTlZgiCVJqzZMwsq6/193nHosrMTXkUp4CXESAwSEU94hu6BD+IUc9BbhIgpSaHRLGScz+Xre7c+BQ+OuQVL2crNKpLP2KRAQYHAPvc2htzoR+/GqXvNUQSnkKcJEUCOdCnuBYOFPGtYz8gHnisEBW2ijARQTIOYnZV2bHGlUd38r7shTgIikQRue2L+jG97nHYvii8RXEnwJcRIA6rL9d9aX0ivByFOAiKRDGGHhLJhsHfXW4cbKEQwEuIsDgcrJ9EeV3tf1p9b/LU4CLpICHMBO8fwjF3WNxGXscaog7BbhIgoV5svGdx2ZvaxuXERTld3kKcJEEC6Pn3S93LZQoVHtSUicxy1OAi6RAGD3xwQt5IloPXHkcOgW4SAw1YhjDou6BR7z/aKQAF0mBMIZSor4Ss+oKleBlKcBFBBgcQoF4DHfE4WrQuFOAi6RAmHelj80QivK7LAW4SIKF2UvtD8yoxt+rnoUSTRmpogAXSbCophFGMYVPPfDwKcBFUmCkYbdgxoScAA+hIKkLBbiIkGmynJOY8UhwncQsTwEukgIjHbc2s5x54CEUFAINoZSnABdJgZFmbqZpcBphdJfSR7v/aKQAFxEyZjl3pY/qU3QtZtgU4CIJFtY4cVOT0RSkgW7okBwKcJEEC2saYSZnDNw9muELDaGETwEukgIjzbqmIUMo8eiBK7/LU4CLpMBII7dpyEnMEZcTCvXAy1OAi8iQHnifeyRzsJXH4VOAi6TBCIc9Jo9tGVwLJYRywpBRF7wsBbhIgoXVU/7KB06PfAy82jw+M7hHp5RWNsDNrM3MHjWzJ83saTO7Ndg+1cxWmdmm4Kd+2yJ1FtYslCnjxgxZCyWSWShVftlMGtvC8vOPC7+QFKmkB34QuMjd3wEsBC4zs3OAm4HV7r4AWB08F5GEGrgnZkxmoYDGzcspG+Ce9VbwtCX4x4GlwIpg+wrg6igKFJEKhNBljttaKFJeRWPgZpYxs7XADmCVuz8CzHT3bQDBzxkl3rvczLrNrLu3tzekskVkiBB6zUNuqTbioxWq+kKeCGpIm4oC3N2PuPtCoAM428xOq/QD3P0Od+9y96729vYayxSRqDVFPOujpqMrxYdV1SwUd98NPARcBmw3s9kAwc8dYRcnIpUJY9Qj6gCvhdYEH14ls1DazWxK8HgscDGwEVgJLAt2WwbcG1GNIlJCFPfEzH8c3vEVxmFrrmCf2cAKM8uQDfy73f0+M3sYuNvMrgc2A9dEWKeIFBHmPTGjztdqZ7co8MsrG+Du/hSwqMj2ncDiKIoSkfSpJZCV4cPTlZgiKRBGzmm8OXkU4CIxVO3QSBgDKUPGwGMS5vGoIr4U4CIpEMbFk1EPV9QyD7wlo4gajn47IgkWl55yJWr5gjjmqLHhF5IiCnCRBIlynZLcL4O4nDwcox74sPTbEUmwJE0jrEVzJoZFxYgCXCRBSk3FCyN84xaVZnDJKbMaXUasKcBFUiBGK8CWVMt4/ZhmRdRw9NsRESB+s1CkPAW4SAqEMxYew9UIZVgKcJEEi2oaodYhSQYFuIgA8RviSNIc90ZRgIskWKjTCEM7ktSLAlwkBcLoreYOm0RzSzV9RYRNAS6SAmH2xMNy4+IF/OXSUweeK77DpwAXSYFQFrMa+SGGuGnJCXzs3M6a368Oe3kKcBEB6hCYCuTQKcBFEiR/MatQ74kZw8WsZHgKcBGRhFKAiyRI/kyOJK1GWO3h9UdAeQpwkRQIO3wVnsmgABdJgWTcUk1fC2FTgIukQPxmgRdSfIdPAS6SYKHOQsm9EjMOveUYlBB3CnARAZSXSaQAF0mwRM1C0TdE6BTgIlIgirDV8rDhU4CLCBC/gI1bPXGkABdJgWRMI4z2+KORAlwkhhpxl3kr8TgsjWhT2inARRIkysWsGmXhnCmNLiGxyga4mc0xs1+Y2QYze9rMbgy2TzWzVWa2Kfh5VPTlikhkGjSEcva8qZzVWRgf+fv/3QfPiKCqZKukB34Y+Ky7nwycA9xgZqcANwOr3X0BsDp4LiIRinQxq4gTvNTFQa3NTRpeqVHZAHf3be7+ePD4TWADcAywFFgR7LYCuDqiGkWkDobkax3POLY2VzaSG4urQ2OmqjFwM+sEFgGPADPdfRtkQx6YUeI9y82s28y6e3t7R1iuiCRVqfg97/jpmqFSo4oD3MwmAPcAn3b3Nyp9n7vf4e5d7t7V3t5eS40iUkYYQymR31GtxAfMmz6++P5lnkuFAW5mLWTD+7vu/pNg83Yzmx28PhvYEU2JIlJKZItZhXbU8KiXXqiSWSgG/BOwwd1vz3lpJbAseLwMuDf88kSkXhqZj5WcxFSAF2quYJ/zgI8B68xsbbDt88BtwN1mdj2wGbgmkgpFpC7GtWaY1NbMH5w1J5LjK4DDVzbA3f3/UvrLeXG45YhINcKcRtjanGHtX1xCU5Px3UdeDu24/YYb7ikW7vmzTtJw0VLYdCWmSAqEFW5NTdGF5Eh74OrBF1KAi6RAmD1xSQ4FuEgahJzfUQxXDHdEXYlZm0pOYopITI00aKdPGMOlp84KqZoyqhwDyd9dV2IWUg9cROpiuPitJJsV34UU4CJSoJ6dXc0uqZ0CXCTBknDy8vRjJjOxdfjRWl3IUxsFuIhE6t8+9R7W3Xpp1e9TXpenABdJgVr74Uma/aGhlkIKcJEEyb+lWlTqHZUVncRUfhdQgIskWBi9UgVjcinARRIkrXOhi/1hUTAPvD6lJIoCXEQKxPF7Io41NZoCXCTBkjCNMDxK8HwKcBEpsGf/obp+3nC967M6jwKgtUVxlU9roYikQK2zU0q9a+akttqLqaWOooVkU/3b17+LfW8fYeOrFd+Kd9RQgIskWCizUIocoylGA85tLRnaWjKNLiOW9DeJiBSIIr9LHjM+3xWJowAXkQJx6oH305WYhRTgIgkW1SyUKO6sNtwwffF7YoZfQ9oowEWkiPqmZ5LWZIkTBbhICtQagKVmr0R4b2MJkQJcJEHqtphVBOMXuit9+BTgIgkW1WJWceiBx6CE2FOAiyRIvRaziuMsFI2TF1KAi0ih+OW3FKEAF0mw6KYRKsGTQAEuIgXiEN/5w0X6TimkABdJgZrviVliu3rgyaAAF0mwcBazKhSHWSj5dBKzkAJcRArVMcAr7ezrj4JCZQPczL5lZjvMbH3OtqlmtsrMNgU/j4q2TBGppyiGUKo9Yv7+LRn1N/NV8hv5Z+CyvG03A6vdfQGwOnguIikRRWd3pCMgrc0K8HxlfyPu/itgV97mpcCK4PEK4OpwyxKRSkQ2jTCGg+BjFOAFav2NzHT3bQDBzxnhlSQi1ap9Mavi26OI770HD4/o/ZPaWkKqJD0i/0ozs+Vm1m1m3b29vVF/nEiq1W8xq/CP2dxUXdzk1zBrcn3v05kEtQb4djObDRD83FFqR3e/w9273L2rvb29xo8TkWLCWcyqPsMlU8ePqcvnjCa1BvhKYFnweBlwbzjliMhw6hW2UQyixHFcPekqmUb4feBh4EQz6zGz64HbgCVmtglYEjwXESlpuPw+YdbE+hWSIs3ldnD3D5d4aXHItYhIig03t/zPLzuJjdve4PHNuwe26SbG5WlejkiCRTWNMArDBfjksS1ceKIms1VLAS6SArUGeb1mtUD59VV0qXz1FOAiCRbVMEMUYZrRSczQKcBFpC5KzUIpFevqkZenABeRutAa4+FTgItIXWgEJXwKcJEYavTNC8aPKTvDuGrqgYdPAS6SYP2zT2pezKrE9hNmTqjtgMNQgIdPAS6SIPVbzCr8sNX9GMKnX6lIgoWzmFUIhVRAPfDwKcBFEqR+i1mFr9xiVjMnabnYainARaQuys1CWbrwmCHPE/xdVTcKcBGpi3JDKLplWvX0GxNJsLQsZiW1UYCLJEjos1DqmP+lAlzBXjsFuIjURbFphF+++jTGtxa/aEjrgZenABdJkPxZKKFMI6xTUBbraf+nMzvq8tlppQAXkbpI8hTIuFKAi0joil2Kv3X3/oJt+Zk+Z+rYgcdNSqeywl+xRkRGvX/4yJkF23757I4hz//6/afT1pIZsu0f/6iL+57cxqSxzSyYUXij4wduOp9L/uevwi02wRTgIgk2uJhVvKYTHl8kfPOHUP7wXXML9jlp1iROmjWp5HFPmDmRT14wn3946PmRF5kC+iNFJEHCDup4xb5Uq+E98EOHDtHT08OBAwcaXUqBtrY2Ojo6aGlpaXQpIkUlaTGrjqPGsmf/ofp82CjR8ADv6elh4sSJdHZ2xuostbuzc+dOenp6mDdvXqPLEQGSPZPjW9edxbtve5Ajfer3h6XhQygHDhxg2rRpsfsP08yYNm1aLP8yEEmimZPaePrWSxtdRqo0PMAhvr2KuNYlIgIxCXAREameAlwkQfJnoQxMIwzpeJIsCnCRFFAOj04Nn4WS69Z/e5pnXnkj1GOecvQkvvj7p4Z6TJFGKbWY1UhO19TzTI9OK4VLPXARSZRJYwevyzh6yui+j2aseuDqKYtIOR8+ey679r7NgUNH+Og5xza6nIYaUQ/czC4zs2fN7Dkzuzmsourp61//OgsXLmThwoW88sorjS5HJNa+8oHTG10Ck8e28PkrTuZLS0/jhJmFa66MJjX3wM0sA3wdWAL0AI+Z2Up3fyas4urhhhtu4IYbbmh0GSKJ0KobD8fKSIZQzgaec/cXAMzsB8BSIFEBLhJHV33t10XvYPPTJ7ayfuuegeebdrwFwL88/PLAtiW3/7Liz9l36EhVdeUv/1qr+e3jQznOaDeSAD8G2JLzvAd4V/5OZrYcWA4wd27h8pEiMujW953KF1c+zYmzBocGjrjzQu9eAC4+eQZjcnrB86aP54FntnP5abN44JntnNExmdmTKz+xd8KsiVx5xuyir33ivfP5xi+f59qz5vCDx7L/qy8+eQYfO+dYtu7ez5SxLZw1byobtr3BlHFj+OrqTRw7bRyfumhByc9rbc7wZxcdzyWnzqq4RinNap3Ib2bXAJe6+8eD5x8Dznb3T5V6T1dXl3d3dw/ZtmHDBk4++eSaaqiHuNcnIulnZmvcvSt/+0gGtHqAOTnPOwCdBRQRqZORBPhjwAIzm2dmY4BrgZW1HCiul/PGtS4RERhBgLv7YeBPgZ8DG4C73f3pao/T1tbGzp07YxeW/euBt7WN7gsFRCS+RnQhj7v/O/DvIzlGR0cHPT099Pb2juQwkei/I4+ISBw1/ErMlpYW3fFGRKQGmpUvIpJQCnARkYRSgIuIJFTNF/LU9GFmvcDLZXcsbjrwWojlxJXamS5qZ3o0so3Hunt7/sa6BvhImFl3sSuR0kbtTBe1Mz3i2EYNoYiIJJQCXEQkoZIU4Hc0uoA6UTvTRe1Mj9i1MTFj4CIiMlSSeuAiIpJDAS4iklCJCPAk3zzZzOaY2S/MbIOZPW1mNwbbp5rZKjPbFPw8Kuc9twRtfdbMLs3Z/k4zWxe89lWzIvfcajAzy5jZE2Z2X/A8de00sylm9mMz2xj8ez03be00s5uC/17Xm9n3zawtLW00s2+Z2Q4zW5+zLbS2mVmrmf0w2P6ImXVG1hh3j/U/QAZ4HjgOGAM8CZzS6LqqqH82cGbweCLwO+AU4G+Bm4PtNwN/Ezw+JWhjKzAvaHsmeO1R4FzAgPuByxvdviLt/QzwPeC+4Hnq2gmsAD4ePB4DTElTO8neLvFFYGzw/G7gurS0ETgfOBNYn7MttLYBnwS+ETy+FvhhZG1p9C+zgl/2ucDPc57fAtzS6LpG0J57gSXAs8DsYNts4Nli7SO73vq5wT4bc7Z/GPhmo9uT17YOYDVwEYMBnqp2ApOCcLO87alpJ4P3u51KdsXS+4BLUtbGzrwAD61t/fsEj5vJXr1pUbQjCUMoxW6efEyDahmR4E+pRcAjwEx33wYQ/JwR7FaqvccEj/O3x8nfA58D+nK2pa2dxwG9wF3BUNGdZjaeFLXT3bcC/x3YDGwD9rj7A6SojUWE2baB93j2xjd7gGlRFJ2EAC82Zpa4uY9mNgG4B/i0u78x3K5Ftvkw22PBzK4Cdrj7mkrfUmRb7NtJtkd1JvB/3H0RsJfsn9ylJK6dwfjvUrJDBkcD483so8O9pci2WLexCrW0rW7tTkKAJ/7myWbWQja8v+vuPwk2bzez2cHrs4EdwfZS7e0JHudvj4vzgPeZ2UvAD4CLzOw7pK+dPUCPuz8SPP8x2UBPUzsvBl509153PwT8BHg36WpjvjDbNvAeM2sGJgO7oig6CQEe2s2TGyE4M/1PwAZ3vz3npZXAsuDxMrJj4/3brw3OZM8DFgCPBn/WvWlm5wTH/KOc9zScu9/i7h3u3kn239GD7v5R0tfOV4EtZnZisGkx8Azpaudm4BwzGxfUtpjsfW/T1MZ8YbYt91gfJPv/QjR/eTT6ZEKFJxyuIDt743ngC42up8ra30P2z6engLXBP1eQHRNbDWwKfk7Nec8XgrY+S85Ze6ALWB+89r+J6MRICG2+gMGTmKlrJ7AQ6A7+nf4rcFTa2gncCmwM6vs22VkYqWgj8H2yY/uHyPaWrw+zbUAb8CPgObIzVY6Lqi26lF5EJKGSMIQiIiJFKMBFRBJKAS4iklAKcBGRhFKAi4gkVHOjCxCpFzM7Aqwj+9/9BmCZu+9rbFUitVMPXEaT/e6+0N1PA94GPtHogkRGQgEuo9WvgeODdaD/1cyeMrPfmtkZAGb2XjNbG/zzhJlNbHC9IgUU4DLqBOtTXE52OOVW4Al3PwP4PPAvwW7/DbjB3RcCvwfsb0CpIsNSgMtoMtbM1pK9DH4z2TVq3kP2UnHc/UFgmplNBn4D3G5mfwZM8eyyoCKxopOYMprsD3rUA0rc4svd/TYz+xnZdWt+a2YXu/vGehQpUin1wGW0+xXwEQAzuwB4zd3fMLP57r7O3f+GbI/9pMaVKFKcFrOSUcPM3nL3CXnbpgJ3kb15wT5gubs/ZWZfAy4EjpBdLvY6dz9Y75pFhqMAFxFJKA2hiIgklAJcRCShFOAiIgmlABcRSSgFuIhIQinARUQSSgEuIpJQ/x9pAmKM39tHGAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "occurance_count2.plot(x=\"Pos\", y=\"-\", kind=\"line\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "cd6a5664-6506-4f37-a9be-9720ae242db5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>-</th>\n",
       "      <th>Eremothecium_cymbalariae_Ecym_2202</th>\n",
       "      <th>Eremothecium_gossypii_AGOS_AFL091W</th>\n",
       "      <th>Kazachstania_africana_KAFR_0C02290</th>\n",
       "      <th>Kazachstania_africana_KAFR_0K00320</th>\n",
       "      <th>Kazachstania_barnettii_KABA2_04S02882</th>\n",
       "      <th>Kazachstania_barnettii_KABA2_06S03410</th>\n",
       "      <th>Kazachstania_exigua_HRR25_1</th>\n",
       "      <th>Kazachstania_exigua_HRR25_2</th>\n",
       "      <th>Kazachstania_naganishii_KNAG_0D01130</th>\n",
       "      <th>...</th>\n",
       "      <th>Zygosaccharomyces_bailii_FUGC01000003.1_cds_SJM83735.1_1</th>\n",
       "      <th>Zygosaccharomyces_parabailii_ZPAR0C01310_A</th>\n",
       "      <th>Zygosaccharomyces_parabailii_ZPAR0K01220_B</th>\n",
       "      <th>Zygosaccharomyces_rouxii_ZYRO0G08866g</th>\n",
       "      <th>Zygotorulaspora_mrakii_HG535_0B00830</th>\n",
       "      <th>a</th>\n",
       "      <th>c</th>\n",
       "      <th>g</th>\n",
       "      <th>t</th>\n",
       "      <th>Pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6460</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6461</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>6461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6462</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6463</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6464</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9039</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>9039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9040</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9041</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>9041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9042</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9043</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9043</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1716 rows × 55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        -  Eremothecium_cymbalariae_Ecym_2202  \\\n",
       "6460  0.0                                 0.0   \n",
       "6461  0.0                                 0.0   \n",
       "6462  0.0                                 0.0   \n",
       "6463  0.0                                 0.0   \n",
       "6464  0.0                                 0.0   \n",
       "...   ...                                 ...   \n",
       "9039  1.0                                 0.0   \n",
       "9040  1.0                                 0.0   \n",
       "9041  1.0                                 0.0   \n",
       "9042  1.0                                 0.0   \n",
       "9043  1.0                                 0.0   \n",
       "\n",
       "      Eremothecium_gossypii_AGOS_AFL091W  Kazachstania_africana_KAFR_0C02290  \\\n",
       "6460                                 0.0                                 0.0   \n",
       "6461                                 0.0                                 0.0   \n",
       "6462                                 0.0                                 0.0   \n",
       "6463                                 0.0                                 0.0   \n",
       "6464                                 0.0                                 0.0   \n",
       "...                                  ...                                 ...   \n",
       "9039                                 0.0                                 0.0   \n",
       "9040                                 0.0                                 0.0   \n",
       "9041                                 0.0                                 0.0   \n",
       "9042                                 0.0                                 0.0   \n",
       "9043                                 0.0                                 0.0   \n",
       "\n",
       "      Kazachstania_africana_KAFR_0K00320  \\\n",
       "6460                                 0.0   \n",
       "6461                                 0.0   \n",
       "6462                                 0.0   \n",
       "6463                                 0.0   \n",
       "6464                                 0.0   \n",
       "...                                  ...   \n",
       "9039                                 0.0   \n",
       "9040                                 0.0   \n",
       "9041                                 0.0   \n",
       "9042                                 0.0   \n",
       "9043                                 0.0   \n",
       "\n",
       "      Kazachstania_barnettii_KABA2_04S02882  \\\n",
       "6460                                    0.0   \n",
       "6461                                    0.0   \n",
       "6462                                    0.0   \n",
       "6463                                    0.0   \n",
       "6464                                    0.0   \n",
       "...                                     ...   \n",
       "9039                                    0.0   \n",
       "9040                                    0.0   \n",
       "9041                                    0.0   \n",
       "9042                                    0.0   \n",
       "9043                                    0.0   \n",
       "\n",
       "      Kazachstania_barnettii_KABA2_06S03410  Kazachstania_exigua_HRR25_1  \\\n",
       "6460                                    0.0                          0.0   \n",
       "6461                                    0.0                          0.0   \n",
       "6462                                    0.0                          0.0   \n",
       "6463                                    0.0                          0.0   \n",
       "6464                                    0.0                          0.0   \n",
       "...                                     ...                          ...   \n",
       "9039                                    0.0                          0.0   \n",
       "9040                                    0.0                          0.0   \n",
       "9041                                    0.0                          0.0   \n",
       "9042                                    0.0                          0.0   \n",
       "9043                                    0.0                          0.0   \n",
       "\n",
       "      Kazachstania_exigua_HRR25_2  Kazachstania_naganishii_KNAG_0D01130  ...  \\\n",
       "6460                          0.0                                   0.0  ...   \n",
       "6461                          0.0                                   0.0  ...   \n",
       "6462                          0.0                                   0.0  ...   \n",
       "6463                          0.0                                   0.0  ...   \n",
       "6464                          0.0                                   0.0  ...   \n",
       "...                           ...                                   ...  ...   \n",
       "9039                          0.0                                   0.0  ...   \n",
       "9040                          0.0                                   0.0  ...   \n",
       "9041                          0.0                                   0.0  ...   \n",
       "9042                          0.0                                   0.0  ...   \n",
       "9043                          0.0                                   0.0  ...   \n",
       "\n",
       "      Zygosaccharomyces_bailii_FUGC01000003.1_cds_SJM83735.1_1  \\\n",
       "6460                                                0.0          \n",
       "6461                                                0.0          \n",
       "6462                                                0.0          \n",
       "6463                                                0.0          \n",
       "6464                                                0.0          \n",
       "...                                                 ...          \n",
       "9039                                                0.0          \n",
       "9040                                                0.0          \n",
       "9041                                                0.0          \n",
       "9042                                                0.0          \n",
       "9043                                                0.0          \n",
       "\n",
       "      Zygosaccharomyces_parabailii_ZPAR0C01310_A  \\\n",
       "6460                                         0.0   \n",
       "6461                                         0.0   \n",
       "6462                                         0.0   \n",
       "6463                                         0.0   \n",
       "6464                                         0.0   \n",
       "...                                          ...   \n",
       "9039                                         0.0   \n",
       "9040                                         0.0   \n",
       "9041                                         0.0   \n",
       "9042                                         0.0   \n",
       "9043                                         0.0   \n",
       "\n",
       "      Zygosaccharomyces_parabailii_ZPAR0K01220_B  \\\n",
       "6460                                         0.0   \n",
       "6461                                         0.0   \n",
       "6462                                         0.0   \n",
       "6463                                         0.0   \n",
       "6464                                         0.0   \n",
       "...                                          ...   \n",
       "9039                                         0.0   \n",
       "9040                                         0.0   \n",
       "9041                                         0.0   \n",
       "9042                                         0.0   \n",
       "9043                                         0.0   \n",
       "\n",
       "      Zygosaccharomyces_rouxii_ZYRO0G08866g  \\\n",
       "6460                                    0.0   \n",
       "6461                                    0.0   \n",
       "6462                                    0.0   \n",
       "6463                                    0.0   \n",
       "6464                                    0.0   \n",
       "...                                     ...   \n",
       "9039                                    0.0   \n",
       "9040                                    0.0   \n",
       "9041                                    0.0   \n",
       "9042                                    0.0   \n",
       "9043                                    0.0   \n",
       "\n",
       "      Zygotorulaspora_mrakii_HG535_0B00830     a    c     g     t   Pos  \n",
       "6460                                   0.0  49.0  0.0   0.0   0.0  6460  \n",
       "6461                                   0.0   0.0  0.0   0.0  49.0  6461  \n",
       "6462                                   0.0   0.0  0.0  48.0   1.0  6462  \n",
       "6463                                   0.0   0.0  0.0  49.0   0.0  6463  \n",
       "6464                                   0.0  48.0  0.0   1.0   0.0  6464  \n",
       "...                                    ...   ...  ...   ...   ...   ...  \n",
       "9039                                   0.0   1.0  1.0   0.0  46.0  9039  \n",
       "9040                                   0.0  28.0  1.0  17.0   2.0  9040  \n",
       "9041                                   0.0   0.0  1.0   0.0  47.0  9041  \n",
       "9042                                   0.0  42.0  0.0   6.0   0.0  9042  \n",
       "9043                                   0.0  30.0  0.0  18.0   0.0  9043  \n",
       "\n",
       "[1716 rows x 55 columns]"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "occurance_count2.loc[occurance_count2['-'] < 45]\n",
    "\n",
    "# identified start codon position as position 6499 of alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "473f4666-93fc-4cc2-87fb-a4e7b760fd63",
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating nucleotide sequence based on start codon & translated sequence\n",
    "\n",
    "prot_regions = []\n",
    "dna_regions  = []\n",
    "\n",
    "pos = 6460\n",
    "test = list(msa_d.keys())[:10]              \n",
    "\n",
    "max_p_len = 0\n",
    "\n",
    "for entry in msa_d:\n",
    "    seq = msa_d[entry][pos::].replace(\"-\",\"\").upper()\n",
    "    # Adjust strat position \n",
    "    i = 1\n",
    "    \n",
    "    # if out of frame and/or not having in-frame AtG, find it \n",
    "    if (seq[:3] != \"ATG\"):\n",
    "        correct= 0\n",
    "        for i in range(1,pos+30):\n",
    "            if correct !=1:           \n",
    "                seq = msa_d[entry][pos+30-i::].replace(\"-\",\"\").upper()\n",
    "                if (seq[:3] != \"ATG\"):\n",
    "                    pass\n",
    "                else:\n",
    "                    correct =1\n",
    "            else:\n",
    "                pass\n",
    "                \n",
    "    prot_seq     = translate_dna2aa(seq)\n",
    "    p_len = len(prot_seq.split(\"*\")[0])\n",
    "    if p_len > max_p_len:\n",
    "        max_p_len = p_len\n",
    "    dna_seq      = seq\n",
    "    \n",
    "    dna_regions.append((entry,dna_seq))    \n",
    "    prot_regions.append((entry,prot_seq))\n",
    "\n",
    "\n",
    "for i in range(len(prot_regions)):\n",
    "    #print(prot_regions[i][1])\n",
    "    if len(prot_regions[i][1])>max_p_len:\n",
    "        prot_regions[i] = (prot_regions[i][0],prot_regions[i][1][:max_p_len])\n",
    "        dna_regions[i]  =  (dna_regions[i][0],dna_regions[i][1][:(max_p_len*3)])\n",
    "        \n",
    "    \n",
    "    \n",
    "    \n",
    "make_fasta_from_listoftups(prot_regions,\"HRR25_homologs_aa_trimmed.fna\")\n",
    "make_fasta_from_listoftups(dna_regions,\"HRR25_homologs_nt_trimmed.fna\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f34dd800-a731-4501-9322-01c4ead891bf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "edcf7edd-3344-4931-9d46-d261dc875a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataList = list(dna_regions)\n",
    "df = pd.DataFrame(\n",
    "      dataList, columns=['Name', 'Seq'])\n",
    "\n",
    "df[['Genome', 'Contig']] = df.Name.str.split(\"_Contig:\", expand = True)\n",
    "df[['Contig', 'Pos']] = df.Contig.str.split(\"_Pos:\", expand = True)\n",
    "df[['Pos', 'Strand']] = df.Pos.str.split(\"_Strand:\", expand = True)\n",
    "df.to_csv(\"HRR25_trimed_nt_info.csv\")\n",
    "\n",
    "\n",
    "\n",
    "dataList = list(prot_regions)\n",
    "df = pd.DataFrame(\n",
    "      dataList, columns=['Name', 'Seq'])\n",
    "df[['Genome', 'Contig']] = df.Name.str.split(\"_Contig:\", expand = True)\n",
    "df[['Contig', 'Pos']] = df.Contig.str.split(\"_Pos:\", expand = True)\n",
    "df[['Pos', 'Strand']] = df.Pos.str.split(\"_Strand:\", expand = True)\n",
    "df.to_csv(\"HRR25_trimed_aa_info.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad99f5ab-1050-4a38-b5ec-636567a94e21",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "576af361-e15a-4c54-aa9c-6862d80a8628",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
