{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ed372500-fb27-472e-a92d-469d22cfc2ef",
   "metadata": {},
   "source": [
    "## Data proccessing for phylogenetic analysis\n",
    "\n",
    "Following https://github.com/abacus-gene/paml-tutorial/tree/main/positive-selection/00_data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1306c1f8-3804-4cf3-9619-8f3744ea3086",
   "metadata": {},
   "source": [
    "Input: HRR25_nuc_nonAligned.fasta\n",
    "\n",
    "mafft output nucletotide: HRR25_mafft_translatorx.nt_ali.fasta\n",
    "mafft output aminoacid  : HRR25_mafft_translatorx.aa_ali.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20e5f9c2-d6c4-4f48-b66b-486e280636a9",
   "metadata": {},
   "source": [
    "# Evaluating gaps in the alignment file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dc0fc9f0-829b-4c92-9660-f8717b068c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "import pandas as pd\n",
    "import math\n",
    "\n",
    "def fasta2dict(f):\n",
    "    d = {}\n",
    "    with  open(f,\"r\") as F:\n",
    "        for line in F:\n",
    "            if line[0]==\">\":\n",
    "                name = line[1:].split(\"\\n\")[0].split(\"|\")[0]\n",
    "                d[name] = ''\n",
    "            else:\n",
    "                #print(name)\n",
    "                d[name]+= line.split(\"\\n\")[0]\n",
    "    F.close()\n",
    "    return d\n",
    "\n",
    "\n",
    "def MSAdict2df_AA(d):\n",
    "    \"\"\"\n",
    "    Dictionary of Multiple Sequence Alignment to pandas dataframe for evaluating gap regions.\n",
    "    \"\"\"\n",
    "    n_entries = len(d)\n",
    "    n_chars   = max([len(d[k]) for k in d])\n",
    "\n",
    "    df = pd.DataFrame(data={k:list(d[k]) for k in d} )\n",
    "    \n",
    "    try:\n",
    "        gap = df.apply(pd.Series.value_counts, axis=1).fillna(0)[\"-\"]    \n",
    "    except KeyError:\n",
    "        gap = pd.Series(data=0, index=[i for i in range(n_chars)])\n",
    "        \n",
    "    \n",
    "    df.insert(0,'Position', [i+1 for i in range(n_chars)])\n",
    "    df.insert(1,'GapRatio', [(float(i)/float(n_entries)) for i in gap])\n",
    "    df2 = df.swapaxes(\"index\", \"columns\")\n",
    "    return df2 \n",
    "\n",
    "\n",
    "\n",
    "def MSAdict2df_nt(d):\n",
    "    \"\"\"\n",
    "    Dictionary of Multiple Sequence Alignment to pandas dataframe for evaluating gap regions.\n",
    "    \"\"\"\n",
    "    n_entries = len(d)-1\n",
    "    n_chars   = max([len(d[k]) for k in d])\n",
    "\n",
    "    df = pd.DataFrame(data={k:list(d[k]) for k in d} )\n",
    "    \n",
    "     \n",
    "    try:\n",
    "        gap = df.apply(pd.Series.value_counts, axis=1).fillna(0)[\"-\"]    \n",
    "    except KeyError:\n",
    "        gap = pd.Series(data=0, index=[i for i in range(n_chars)])\n",
    "  \n",
    "    df.insert(0,'Position', [math.floor(i/3)+1 for i in range(n_chars)])\n",
    "    df.insert(1,'GapRatio', [(float(i+0)/float(n_entries)) for i in gap])\n",
    "    df2 = df.swapaxes(\"index\", \"columns\")\n",
    "    return df2 \n",
    "\n",
    "\n",
    "def df2fasta(df,f):\n",
    "    with open(f,\"w\") as F:\n",
    "        for index, row in df.iterrows():\n",
    "            gene = index\n",
    "            seq  = (\"\").join([row[i] for i in df.columns.values.tolist()])\n",
    "            entry = \">%s\\n%s\\n\"%(gene,seq)\n",
    "            F.write(entry)\n",
    "        F.close()\n",
    "    print(\"Generated fasta: %s\"%(f))\n",
    "        \n",
    "    \n",
    "def df2phy(df,f):\n",
    "    nrow  = len(df.index)\n",
    "    nchar = len(df. columns)\n",
    "    with open(f,\"w\") as F:\n",
    "        F.write(\"%d   %d\\n\\n\"%(nrow,nchar))\n",
    "        \n",
    "        for index, row in df.iterrows():\n",
    "            gene = index\n",
    "            seq  = (\"\").join([row[i] for i in df.columns.values.tolist()])\n",
    "            entry = \"%s      %s\\n\"%(gene,seq)\n",
    "            F.write(entry)\n",
    "        F.close() \n",
    "        \n",
    "    \n",
    "def removekey(d, key):\n",
    "    r = dict(d)\n",
    "    del r[key]\n",
    "    return r\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "988576d2-2300-44a3-b999-8f037c84d3a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>2684</th>\n",
       "      <th>2685</th>\n",
       "      <th>2686</th>\n",
       "      <th>2687</th>\n",
       "      <th>2688</th>\n",
       "      <th>2689</th>\n",
       "      <th>2690</th>\n",
       "      <th>2691</th>\n",
       "      <th>2692</th>\n",
       "      <th>2693</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Position</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>895</td>\n",
       "      <td>896</td>\n",
       "      <td>896</td>\n",
       "      <td>896</td>\n",
       "      <td>897</td>\n",
       "      <td>897</td>\n",
       "      <td>897</td>\n",
       "      <td>898</td>\n",
       "      <td>898</td>\n",
       "      <td>898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GapRatio</th>\n",
       "      <td>0.011364</td>\n",
       "      <td>0.011364</td>\n",
       "      <td>0.011364</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.977273</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.136364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HRR25_19.1</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HRR25_19.2</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HRR25_18.1</th>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2694 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                0         1         2    3    4    5    6    7    8    9     \\\n",
       "Position           1         1         1    2    2    2    3    3    3    4   \n",
       "GapRatio    0.011364  0.011364  0.011364  1.0  1.0  1.0  1.0  1.0  1.0  1.0   \n",
       "HRR25_19.1         A         T         G    -    -    -    -    -    -    -   \n",
       "HRR25_19.2         A         T         G    -    -    -    -    -    -    -   \n",
       "HRR25_18.1         A         T         G    -    -    -    -    -    -    -   \n",
       "\n",
       "            ...      2684      2685      2686      2687      2688      2689  \\\n",
       "Position    ...       895       896       896       896       897       897   \n",
       "GapRatio    ...  0.977273  0.977273  0.977273  0.977273  0.136364  0.136364   \n",
       "HRR25_19.1  ...         C         C         T         G         T         G   \n",
       "HRR25_19.2  ...         -         -         -         -         T         G   \n",
       "HRR25_18.1  ...         A         T         T         A         C         A   \n",
       "\n",
       "                2690      2691      2692      2693  \n",
       "Position         897       898       898       898  \n",
       "GapRatio    0.136364  0.136364  0.136364  0.136364  \n",
       "HRR25_19.1         G         C         T         C  \n",
       "HRR25_19.2         G         C         T         A  \n",
       "HRR25_18.1         A         T         T         T  \n",
       "\n",
       "[5 rows x 2694 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nt_f = \"./raw_data/HRR25_mafft_translatorx.nt_ali.fasta\"\n",
    "aa_f = \"./raw_data/HRR25_mafft_translatorx.aa_ali.fasta\"\n",
    "\n",
    "aa_d = fasta2dict(aa_f)\n",
    "nt_d = fasta2dict(nt_f)\n",
    "\n",
    "\n",
    "# removing post-WGD species with aleged low quality\n",
    "remove_nodes = ['HRR25_04.1','HRR25_02.1','HRR25_01.1','HRR25_03.1','HRR25_07.1']\n",
    "\n",
    "for k in remove_nodes:\n",
    "    aa_d = removekey(aa_d, k)\n",
    "    nt_d = removekey(nt_d, k)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "df_aa= MSAdict2df_AA(aa_d)\n",
    "df_aa.head()\n",
    "\n",
    "df_nt= MSAdict2df_nt(nt_d)\n",
    "df_nt.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f07f5fa-e665-48c4-9ba2-abefd3857465",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c6539c9f-bd9c-4f32-9678-fb24546dade1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>888</th>\n",
       "      <th>889</th>\n",
       "      <th>890</th>\n",
       "      <th>891</th>\n",
       "      <th>892</th>\n",
       "      <th>893</th>\n",
       "      <th>894</th>\n",
       "      <th>895</th>\n",
       "      <th>896</th>\n",
       "      <th>897</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Position</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>889</td>\n",
       "      <td>890</td>\n",
       "      <td>891</td>\n",
       "      <td>892</td>\n",
       "      <td>893</td>\n",
       "      <td>894</td>\n",
       "      <td>895</td>\n",
       "      <td>896</td>\n",
       "      <td>897</td>\n",
       "      <td>898</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GapRatio</th>\n",
       "      <td>0.011236</td>\n",
       "      <td>0.988764</td>\n",
       "      <td>0.988764</td>\n",
       "      <td>0.988764</td>\n",
       "      <td>0.011236</td>\n",
       "      <td>0.011236</td>\n",
       "      <td>0.011236</td>\n",
       "      <td>0.011236</td>\n",
       "      <td>0.011236</td>\n",
       "      <td>0.011236</td>\n",
       "      <td>...</td>\n",
       "      <td>0.977528</td>\n",
       "      <td>0.977528</td>\n",
       "      <td>0.977528</td>\n",
       "      <td>0.977528</td>\n",
       "      <td>0.988764</td>\n",
       "      <td>0.977528</td>\n",
       "      <td>0.966292</td>\n",
       "      <td>0.966292</td>\n",
       "      <td>0.134831</td>\n",
       "      <td>0.134831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HRR25_19.1</th>\n",
       "      <td>M</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>D</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "      <td>V</td>\n",
       "      <td>G</td>\n",
       "      <td>R</td>\n",
       "      <td>...</td>\n",
       "      <td>H</td>\n",
       "      <td>E</td>\n",
       "      <td>Q</td>\n",
       "      <td>I</td>\n",
       "      <td>E</td>\n",
       "      <td>G</td>\n",
       "      <td>P</td>\n",
       "      <td>L</td>\n",
       "      <td>W</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HRR25_19.2</th>\n",
       "      <td>M</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>D</td>\n",
       "      <td>L</td>\n",
       "      <td>R</td>\n",
       "      <td>V</td>\n",
       "      <td>G</td>\n",
       "      <td>R</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>W</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HRR25_18.1</th>\n",
       "      <td>M</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>E</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "      <td>V</td>\n",
       "      <td>G</td>\n",
       "      <td>K</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>K</td>\n",
       "      <td>L</td>\n",
       "      <td>Q</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 898 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0         1         2         3         4         5    \\\n",
       "Position           1         2         3         4         5         6   \n",
       "GapRatio    0.011236  0.988764  0.988764  0.988764  0.011236  0.011236   \n",
       "HRR25_19.1         M         -         -         -         D         L   \n",
       "HRR25_19.2         M         -         -         -         D         L   \n",
       "HRR25_18.1         M         -         -         -         E         V   \n",
       "\n",
       "                 6         7         8         9    ...       888       889  \\\n",
       "Position           7         8         9        10  ...       889       890   \n",
       "GapRatio    0.011236  0.011236  0.011236  0.011236  ...  0.977528  0.977528   \n",
       "HRR25_19.1         R         V         G         R  ...         H         E   \n",
       "HRR25_19.2         R         V         G         R  ...         -         -   \n",
       "HRR25_18.1         K         V         G         K  ...         -         -   \n",
       "\n",
       "                 890       891       892       893       894       895  \\\n",
       "Position         891       892       893       894       895       896   \n",
       "GapRatio    0.977528  0.977528  0.988764  0.977528  0.966292  0.966292   \n",
       "HRR25_19.1         Q         I         E         G         P         L   \n",
       "HRR25_19.2         -         -         -         -         -         -   \n",
       "HRR25_18.1         -         -         -         -         K         L   \n",
       "\n",
       "                 896       897  \n",
       "Position         897       898  \n",
       "GapRatio    0.134831  0.134831  \n",
       "HRR25_19.1         W         L  \n",
       "HRR25_19.2         W         L  \n",
       "HRR25_18.1         Q         F  \n",
       "\n",
       "[5 rows x 898 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_aa.to_csv(\"./data_prep4codeml/HRR25_mafft_gap_evaluation_aa.csv\", sep=',')\n",
    "df_nt.to_csv(\"./data_prep4codeml/HRR25_mafft_gap_evaluation_nt.csv\", sep=',')\n",
    "\n",
    "\n",
    "\n",
    "df_aa.head()\n",
    "#df_aa = df_aa.drop([\"Name\"], axis=1)\n",
    "#df_nt = df_nt.drop([\"Name\"], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "96b54b97-fc1e-4a70-acd8-7639a81d5f69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated fasta: ./data_prep4codeml/HRR25_mafft_translatorx_aa_ali_nogaps.fasta\n",
      "Generated fasta: ./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps.fasta\n"
     ]
    }
   ],
   "source": [
    "gap_ratio = 0.10\n",
    "# Define protein positions to be analyzed downstream. Here I take protein position which has ≥90% occupancy\n",
    "Positions = list(df_aa.loc[\"Position\"][df_aa.loc['GapRatio'] <= gap_ratio])\n",
    "# Defining column index of protein and nucleotide to extract\n",
    "AA_col    = [i for i in df_aa.columns if df_aa[i][\"Position\"] in Positions]\n",
    "nt_col    = [i for i in df_nt.columns if df_nt[i][\"Position\"] in Positions]\n",
    "\n",
    "# Extracting sequence based on protein potision defined above\n",
    "df_aa_nogaps = df_aa[df_aa.columns[df_aa.columns.isin(AA_col)]]\n",
    "df_nt_nogaps = df_nt[df_nt.columns[df_nt.columns.isin(nt_col)]]\n",
    "\n",
    "#df_aa_nogaps = df_aa_nogaps.drop_duplicates(keep='last')\n",
    "#df_nt_nogaps = df_nt_nogaps.drop_duplicates(keep='last')\n",
    "df2fasta(df_aa_nogaps.drop([\"Position\", \"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_aa_ali_nogaps.fasta\")\n",
    "df2fasta(df_nt_nogaps.drop([\"Position\", \"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps.fasta\")\n",
    "\n",
    "df2phy(df_nt_nogaps.drop([\"Position\",\"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps.phy\")\n",
    "df2phy(df_nt_nogaps.iloc[:, [i for i in range(24)]].drop([\"Position\",\"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps_4test.phy\")\n",
    "\n",
    "df_aa_nogaps.drop([\"Position\",\"GapRatio\"]).to_csv(\"./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps.csv\", sep=',')\n",
    "df_aa_nogaps.drop([\"Position\",\"GapRatio\"]).to_csv(\"./data_prep4codeml/HRR25_mafft_translatorx_aa_ali_nogaps.csv\", sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "97ed8ffd-cb53-4ca7-ac59-7d28fc969d87",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref = \"HRR25_20.1\"\n",
    "\n",
    "pos_domains = {}\n",
    "\n",
    "pos = 0\n",
    "for i in range(0,len(aa_d[ref])):\n",
    "    msa_pos = i+1\n",
    "    if aa_d[ref][i] != '-':\n",
    "        pos+=1\n",
    "        if pos < 291:\n",
    "            domain = \"kinase\"\n",
    "        elif pos < 395:\n",
    "            domain = \"central\"\n",
    "        else:\n",
    "            domain = \"tail\"\n",
    "        pos_domains[msa_pos] = domain\n",
    "    else:\n",
    "        if pos < 291:\n",
    "            domain = \"kinase\"\n",
    "        elif pos < 395:\n",
    "            domain = \"central\"\n",
    "        else:\n",
    "            domain = \"tail\"\n",
    "        pos_domains[msa_pos] = domain\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f1e21724-573b-4425-a78c-a03d205674db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Generated fasta: ./data_prep4codeml/HRR25_mafft_translatorx_aa_ali_nogaps_kinasedomain.fasta\n",
      "Generated fasta: ./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps_kinasedomain.fasta\n",
      "Generated fasta: ./data_prep4codeml/HRR25_mafft_translatorx_aa_ali_nogaps_centraldomain.fasta\n",
      "Generated fasta: ./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps_centraldomain.fasta\n",
      "Generated fasta: ./data_prep4codeml/HRR25_mafft_translatorx_aa_ali_nogaps_taildomain.fasta\n",
      "Generated fasta: ./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps_taildomain.fasta\n"
     ]
    }
   ],
   "source": [
    "gap_ratio = 0.10\n",
    "# Define protein positions to be analyzed downstream. Here I take protein position which has ≥90% occupancy\n",
    "Positions = list(df_aa.loc[\"Position\"][df_aa.loc['GapRatio'] <= gap_ratio])\n",
    "Positions_domain = [i for i in Positions if pos_domains[i]==\"kinase\"] \n",
    "# Defining column index of protein and nucleotide to extract\n",
    "AA_col_domain    = [i for i in df_aa.columns if df_aa[i][\"Position\"] in Positions_domain]\n",
    "nt_col_domain    = [i for i in df_nt.columns if df_nt[i][\"Position\"] in Positions_domain]\n",
    "# Extracting sequence based on protein potision defined above\n",
    "df_aa_nogaps_domain = df_aa[df_aa.columns[df_aa.columns.isin(AA_col_domain)]]\n",
    "df_nt_nogaps_domain = df_nt[df_nt.columns[df_nt.columns.isin(nt_col_domain)]]\n",
    "df2fasta(df_aa_nogaps_domain.drop([\"Position\", \"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_aa_ali_nogaps_kinasedomain.fasta\")\n",
    "df2fasta(df_nt_nogaps_domain.drop([\"Position\", \"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps_kinasedomain.fasta\")\n",
    "df2phy(df_nt_nogaps_domain.drop([\"Position\",\"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps_kinasedomain.phy\")\n",
    "\n",
    "\n",
    "Positions_domain = [i for i in Positions if pos_domains[i]==\"central\"] \n",
    "# Defining column index of protein and nucleotide to extract\n",
    "AA_col_domain    = [i for i in df_aa.columns if df_aa[i][\"Position\"] in Positions_domain]\n",
    "nt_col_domain    = [i for i in df_nt.columns if df_nt[i][\"Position\"] in Positions_domain]\n",
    "# Extracting sequence based on protein potision defined above\n",
    "df_aa_nogaps_domain = df_aa[df_aa.columns[df_aa.columns.isin(AA_col_domain)]]\n",
    "df_nt_nogaps_domain = df_nt[df_nt.columns[df_nt.columns.isin(nt_col_domain)]]\n",
    "df2fasta(df_aa_nogaps_domain.drop([\"Position\", \"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_aa_ali_nogaps_centraldomain.fasta\")\n",
    "df2fasta(df_nt_nogaps_domain.drop([\"Position\", \"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps_centraldomain.fasta\")\n",
    "df2phy(df_nt_nogaps_domain.drop([\"Position\",\"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps_centraldomain.phy\")\n",
    "\n",
    "\n",
    "Positions_domain = [i for i in Positions if pos_domains[i]==\"tail\"] \n",
    "# Defining column index of protein and nucleotide to extract\n",
    "AA_col_domain    = [i for i in df_aa.columns if df_aa[i][\"Position\"] in Positions_domain]\n",
    "nt_col_domain    = [i for i in df_nt.columns if df_nt[i][\"Position\"] in Positions_domain]\n",
    "# Extracting sequence based on protein potision defined above\n",
    "df_aa_nogaps_domain = df_aa[df_aa.columns[df_aa.columns.isin(AA_col_domain)]]\n",
    "df_nt_nogaps_domain = df_nt[df_nt.columns[df_nt.columns.isin(nt_col_domain)]]\n",
    "df2fasta(df_aa_nogaps_domain.drop([\"Position\", \"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_aa_ali_nogaps_taildomain.fasta\")\n",
    "df2fasta(df_nt_nogaps_domain.drop([\"Position\", \"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps_taildomain.fasta\")\n",
    "df2phy(df_nt_nogaps_domain.drop([\"Position\",\"GapRatio\"]),\"./data_prep4codeml/HRR25_mafft_translatorx_nt_ali_nogaps_taildomain.phy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f03d9c72-3af8-4a1c-8489-bcf807486304",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
