{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e93f8aa4-7960-4ea6-ba91-9ba7ab48bf71",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import re\n",
    "\n",
    "from pprint import pprint\n",
    "from itertools import groupby\n",
    "import pandas as pd\n",
    "\n",
    "###############\n",
    "### Small functions\n",
    "###############\n",
    "def rev_comp(seq):\n",
    "    complement_dict = {'A':'T','T':'A','G':'C','C':'G','N':'N'}\n",
    "    return \"\".join([complement_dict[base.upper()] for base in reversed(seq)])\n",
    "\n",
    "def translate_dna2aa(dna_seq):\n",
    "    dna_seq = dna_seq.upper()\n",
    "    codon_table = {\n",
    "    'GCT':\"A\",\n",
    "    'GCC':\"A\",\n",
    "    'GCA':\"A\",\n",
    "    'GCG':\"A\",\n",
    "    'TGT':\"C\",\n",
    "    'TGC':\"C\",\n",
    "    'GAT':\"D\",\n",
    "    'GAC':\"D\",\n",
    "    'GAA':\"E\",\n",
    "    'GAG':\"E\",\n",
    "    'TTT':'F',\n",
    "    'TTC':'F',\n",
    "    'GGT':'G',\n",
    "    'GGC':'G',\n",
    "    'GGA':'G',\n",
    "    'GGG':'G',\n",
    "    'CAT':'H',\n",
    "    'CAC':'H',\n",
    "    'ATT':'I',\n",
    "    'ATC':'I',\n",
    "    'ATA':'I',\n",
    "    'AAA':'K',\n",
    "    'AAG':'K',\n",
    "    'TTA':'L',\n",
    "    'TTG':'L',\n",
    "    'CTT':'L',\n",
    "    'CTC':'L',\n",
    "    'CTA':'L',\n",
    "    'CTG':'L',\n",
    "    'ATG':'M',\n",
    "    'AAT':'N', 'AAC':'N',\n",
    "    'CCT':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P',\n",
    "    'CAA':'Q', 'CAG':'Q',\n",
    "    'CGT':'R', 'CGC':'R', 'CGA':'R', 'CGG':'R', 'AGA':'R', 'AGG':'R',\n",
    "    'TCT':'S', 'TCC':'S', 'TCA':'S', 'TCG':'S', 'AGT':'S', 'AGC':'S',\n",
    "    'ACT':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T',\n",
    "    'GTT':'V', 'GTC':'V', 'GTA':'V', 'GTG':'V',\n",
    "    'TGG':'W',\n",
    "    'TAT':'Y', 'TAC':'Y',\n",
    "    'TAG':\"*\",'TGA':\"*\",'TAA':\"*\",\n",
    "\n",
    "    #Inverse table for the standard genetic code (compressed using IUPAC notation)\n",
    "    'GCN':\"A\",\n",
    "    'CGN':\"R\",'AGR':\"R\",'CGY':\"R\",'MGR':\"R\",\n",
    "    'AAY':\"N\",\n",
    "    'GAY':\"D\",\n",
    "    'RAY':\"B\",\n",
    "    'TGY':\"C\",\n",
    "    'CAR':\"Q\",\n",
    "    'GAR':\"E\",\n",
    "    'SAR':\"Z\",\n",
    "    'GGN':\"G\",\n",
    "    'CAY':\"H\",\n",
    "    'ATH':\"I\",\n",
    "    'CTN':\"L\",'TTR':\"L\",'CTY':\"L\",'YTR':\"L\",\n",
    "    'AAR':\"K\",\n",
    "    'TTY':\"F\",\n",
    "    'CCN':\"P\",\n",
    "    'TCN':\"S\",'AGY':\"S\",\n",
    "    'ACN':\"T\",\n",
    "    'TAY':\"Y\",\n",
    "    'GTN':\"V\",\n",
    "    'TRA':\"*\",'TAR':\"*\"\n",
    "                }\n",
    "\n",
    "    aa_seq = \"\"\n",
    "    for i in range(1,int(len(dna_seq)/3)+1):\n",
    "        try:\n",
    "            aa_seq += codon_table[dna_seq[(i-1)*3:i*3]]\n",
    "        except KeyError:\n",
    "            #print(\"Undefined codon: '%s' was translated to X \"%(dna_seq[(i-1)*3:i*3]))\n",
    "            aa_seq += 'X'\n",
    "    return(aa_seq)\n",
    "\n",
    "def fasta2dict(f):\n",
    "    d = {}\n",
    "    with  open(f,\"r\") as F:\n",
    "        for line in F:\n",
    "            if line[0]==\">\":\n",
    "                name = line[1:].split(\"\\n\")[0].split(\"|\")[0]\n",
    "                d[name] = ''\n",
    "            else:\n",
    "                #print(name)\n",
    "                d[name]+= line.split(\"\\n\")[0]\n",
    "    F.close()\n",
    "\n",
    "    return d\n",
    "\n",
    "        \n",
    "\n",
    "\n",
    "def make_fasta_from_listoftups(L,name):\n",
    "    with open(\"%s\"%name,\"w\") as f:\n",
    "        for I in L:\n",
    "            #print I\n",
    "            f.write('>%s\\n'%(str(I[0])))\n",
    "            f.write('%s\\n'%(str(I[1])))\n",
    "        f.close()\n",
    "        #print(\"Made fna file : %s\"%(name))\n",
    "\n",
    "\n",
    "def get_blastout(x):\n",
    "    files = []\n",
    "    lis = os.listdir(\"%s\" %x)\n",
    "    for i in lis:\n",
    "        if i[-4:] == \".out\" :\n",
    "            files.append(i)\n",
    "    return list(files)\n",
    "\n",
    "def parse_blastout(dir_out):\n",
    "    lis = os.listdir(dir_out)\n",
    "    fs = [\"%s/%s\" %(dir_out,f) for f in lis]\n",
    "    D = {}\n",
    "    for f in fs:\n",
    "        if f.split(\"/\")[-1][0] != \".\":\n",
    "            genome = f.split(\"/\")[-1].split(\".blast\")[0]\n",
    "            with open(f, \"r\") as F:\n",
    "                hit_n = 0\n",
    "                for line in F:\n",
    "                    \"\"\"\n",
    "                    'qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore', \n",
    "                    \"\"\"\n",
    "                    hit_n+=1\n",
    "                    hit         = hit_n   \n",
    "                    cols        =    line.split(\",\")\n",
    "                    contig      =    cols[0]\n",
    "                    subject     =    cols[1]\n",
    "                    pident      =    float(cols[2])\n",
    "                    str_on_q    =    int(cols[6])\n",
    "                    end_on_q    =    int(cols[7])\n",
    "                    str_on_s    =    int(cols[8])\n",
    "                    end_on_s    =    int(cols[9])\n",
    "                    evalue      =    float(cols[10])\n",
    "\n",
    "\n",
    "                    el = {}\n",
    "                    el[\"contig\"]   = contig\n",
    "                    el[\"subject\"]    = subject\n",
    "                    el[\"str_s\"]      = str_on_s\n",
    "                    el[\"end_s\"]      = end_on_s\n",
    "                    el[\"str_q\"]      = str_on_q\n",
    "                    el[\"end_q\"]      = end_on_q\n",
    "                    el[\"evalue\"]   = evalue\n",
    "                    el[\"pident\"] = pident\n",
    "                    try:\n",
    "                        D[genome][hit] = el\n",
    "                    except KeyError:        \n",
    "                        D[genome] = {}\n",
    "                        D[genome][hit] = el\n",
    "\n",
    "            F.close()\n",
    "    return D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "876c7415-4942-4a3a-950e-733f8d0230a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "dir_genomes = \"./genomes\"\n",
    "db          = \"./blast/db/HRR25_geneanalysis_nt_translatorXinput.fna\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "58443e72-db67-4f98-a96f-c3e835c15469",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_dir     =  \"./blast_/out\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2dfc1e4e-64b8-4ed0-a51b-bb8322853111",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_fs = os.listdir(dir_genomes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ff21338b-cb48-4a4f-b530-71604f567eb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 out of 68 files. Curently working on GCA_000292725.1_SacArb1.0_genomic.fna\n",
      "2 out of 68 files. Curently working on GCA_030580015.1_ASM3058001v1_genomic.fna\n",
      "3 out of 68 files. Curently working on GCA_030557885.1_ASM3055788v1_genomic.fna\n",
      "4 out of 68 files. Curently working on GCF_001298625.1_SEUB3.0_genomic.fna\n",
      "5 out of 68 files. Curently working on yHAB133_kazachstania_unispora_160519.fas\n",
      "6 out of 68 files. Curently working on GCA_000710315.1_Eremothecium_coryli_genomic.fna\n",
      "7 out of 68 files. Curently working on yHAB160_kazachstania_kunashirensis_160519.fas\n",
      "8 out of 68 files. Curently working on .DS_Store\n",
      "9 out of 68 files. Curently working on yHAB154_kazachstania_transvaalensis_160519.fas\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Error: NCBI C++ Exception:\n",
      "    T0 \"/Users/coremake/release_build/build/PrepareRelease_IntelMAC-Clang36_JSID_01_90273_130.14.22.10_9008__PrepareRelease_IntelMAC-Clang36_1481139955/c++/compilers/unix/../../src/objtools/readers/fasta.cpp\", line 2428: Error: CFastaReader: Near line 1, there's a line that doesn't look like plausible data, but it's not marked as defline or comment. (m_Pos = 1)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10 out of 68 files. Curently working on GCF_947241705.1_Smik-IFO1815_genomic.fna\n",
      "11 out of 68 files. Curently working on yHAB153_kazachstania_rosinii_160519.fas\n",
      "12 out of 68 files. Curently working on GCA_017309295.1_JHU_Cniv_v1_genomic.fna\n",
      "13 out of 68 files. Curently working on yHMPu5000034877_tetrapisispora_namnaonensis_160519.fas\n",
      "14 out of 68 files. Curently working on GCA_030556345.1_ASM3055634v1_genomic.fna\n",
      "15 out of 68 files. Curently working on GCF_001417885.1_Kmar_1.0_genomic.fna\n",
      "16 out of 68 files. Curently working on yHMPu5000034881_torulaspora_pretoriensis_160519.fas\n",
      "17 out of 68 files. Curently working on tetrapisispora_phaffii.fas\n",
      "18 out of 68 files. Curently working on GCA_030569995.1_ASM3056999v1_genomic.fna\n",
      "19 out of 68 files. Curently working on yHMPu5000026152_torulaspora_franciscae_160519.fas\n",
      "20 out of 68 files. Curently working on GCA_900074735.1_LAFA0_genomic.fna\n",
      "21 out of 68 files. Curently working on yHAB159_kazachstania_solicola_160519.fas\n",
      "22 out of 68 files. Curently working on yHAB164_kazachstania_aerobia_160519.fas\n",
      "23 out of 68 files. Curently working on ashbya_aceri.fas\n",
      "24 out of 68 files. Curently working on yHAB136_kazachstania_bromeliacearum_160519.fas\n",
      "25 out of 68 files. Curently working on yHMPu5000034865_zygosaccharomyces_kombuchaensis_160519.fas\n",
      "26 out of 68 files. Curently working on GCF_000938715.1_LALA0_genomic.fna\n",
      "27 out of 68 files. Curently working on GCF_000150035.1_ASM15003v1_genomic.fna\n",
      "28 out of 68 files. Curently working on GCF_947243775.1_Skud-IFO1802_genomic.fna\n",
      "29 out of 68 files. Curently working on yHMPu5000034710_kluyveromyces_dobzhanskii_160519.fas\n",
      "30 out of 68 files. Curently working on GCA_900074755.1_LANO0_genomic.fna\n",
      "31 out of 68 files. Curently working on yHAB132_kazachstania_martiniae_160519.fas\n",
      "32 out of 68 files. Curently working on GCA_030569875.1_ASM3056987v1_genomic.fna\n",
      "33 out of 68 files. Curently working on yHAB148_kazachstania_intestinalis_160519.fas\n",
      "34 out of 68 files. Curently working on GCA_900074715.1_LAME0_genomic.fna\n",
      "35 out of 68 files. Curently working on GCA_949129075.1_ZBA_7846_DN_genomic.fna\n",
      "36 out of 68 files. Curently working on yHAB155_kazachstania_spencerorum_160519.fas\n",
      "37 out of 68 files. Curently working on GCF_000235365.1_ASM23536v1_genomic.fna\n",
      "38 out of 68 files. Curently working on yHMPu5000035652_torulaspora_maleeae_160613.fas\n",
      "39 out of 68 files. Curently working on GCA_900074765.1_LAFE0_genomic.fna\n",
      "40 out of 68 files. Curently working on GCF_001548555.1_ASM154855v1_genomic.fna\n",
      "41 out of 68 files. Curently working on GCA_030573655.1_ASM3057365v1_genomic.fna\n",
      "42 out of 68 files. Curently working on GCF_000142805.1_ASM14280v1_genomic.fna\n",
      "43 out of 68 files. Curently working on yHMPu5000034862_zygotorulaspora_florentina_160519.fas\n",
      "44 out of 68 files. Curently working on GCF_000002515.2_ASM251v1_genomic.fna\n",
      "45 out of 68 files. Curently working on yHAB143_kazachstania_siamensis_160519.fas\n",
      "46 out of 68 files. Curently working on yHAB158_kazachstania_viticola_160519.fas\n",
      "47 out of 68 files. Curently working on GCF_000292725.1_SacArb1.0_genomic.fna\n",
      "48 out of 68 files. Curently working on GCA_001298625.1_SEUB3.0_genomic.fna\n",
      "49 out of 68 files. Curently working on GCF_000091025.4_ASM9102v4_genomic.fna\n",
      "50 out of 68 files. Curently working on yHMPu5000034709_kluyveromyces_aestuarii_160519.fas\n",
      "51 out of 68 files. Curently working on GCA_001039675.1_AGH_PRJEB150_V1_genomic.fna\n",
      "52 out of 68 files. Curently working on yHAB134_kazachstania_turicensis_160519.fas\n",
      "53 out of 68 files. Curently working on GCA_900074745.1_LAMI0_genomic.fna\n",
      "54 out of 68 files. Curently working on GCA_947243795.1_Suva-ZP964_genomic.fna\n",
      "55 out of 68 files. Curently working on yHMPu5000034866_zygosaccharomyces_bisporus_160519.fas\n",
      "56 out of 68 files. Curently working on GCF_000026365.1_ASM2636v1_genomic.fna\n",
      "57 out of 68 files. Curently working on yHAB166_kazachstania_yakushimaensis_160519.fas\n",
      "58 out of 68 files. Curently working on GCA_000149225.2_ASM14922v2_genomic.fna\n",
      "59 out of 68 files. Curently working on GCA_900474375.2_BTMP13_1_contigs_genomic.fna\n",
      "60 out of 68 files. Curently working on yHMPu5000034875_tetrapisispora_fleetii_160519.fas\n",
      "61 out of 68 files. Curently working on GCF_000002545.3_ASM254v2_genomic.fna\n",
      "62 out of 68 files. Curently working on GCA_900074725.1_LADA0_genomic.fna\n",
      "63 out of 68 files. Curently working on yHMPu5000034876_tetrapisispora_iriomotensis_160519.fas\n",
      "64 out of 68 files. Curently working on yHMPu5000035651_torulaspora_microellipsoides_160519.fas\n",
      "65 out of 68 files. Curently working on GCF_002079055.1_ASM207905v1_genomic.fna\n",
      "66 out of 68 files. Curently working on yHMPu5000035629_yueomyces_sinensis_160519.fas\n",
      "67 out of 68 files. Curently working on yHMPu5000026256_zygotorulaspora_mrakii_161215.fas\n",
      "68 out of 68 files. Curently working on yHAB147_kazachstania_taianensis_160519.fas\n"
     ]
    }
   ],
   "source": [
    "# For each genome, blast the orthologs from YGOB\n",
    "\n",
    "i = 0\n",
    "for f in genome_fs:\n",
    "    i +=1\n",
    "    print(\"%d out of %d files. Curently working on %s\"%(i,len(genome_fs),f))\n",
    "    nm = f\n",
    "    command = \"blastn -task blastn -query %s/%s -db %s -outfmt 10 -evalue 1e-20 > %s/%s.blast\"%(dir_genomes,f,db, out_dir,nm)\n",
    "    os.system(command)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "76fb85a4-10bc-4c8e-b37f-f703ab603657",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge blast result to dictionary\n",
    "blast_d = parse_blastout(out_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "06f41dda-ef48-4d13-a1c4-b4ad903e4fd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>Contig</th>\n",
       "      <th>q_start</th>\n",
       "      <th>q_end</th>\n",
       "      <th>Subject</th>\n",
       "      <th>s_start</th>\n",
       "      <th>s_end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_030569995.1_ASM3056999v1_genomic.fna</td>\n",
       "      <td>JAJLUF010000032.1</td>\n",
       "      <td>8108</td>\n",
       "      <td>9496</td>\n",
       "      <td>HRR25_60.1</td>\n",
       "      <td>1389</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tetrapisispora_phaffii.fas</td>\n",
       "      <td>NC_016524.1</td>\n",
       "      <td>591805</td>\n",
       "      <td>593361</td>\n",
       "      <td>HRR25_35.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1557</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tetrapisispora_phaffii.fas</td>\n",
       "      <td>NC_016529.1</td>\n",
       "      <td>523631</td>\n",
       "      <td>525016</td>\n",
       "      <td>HRR25_35.2</td>\n",
       "      <td>1386</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_900074765.1_LAFE0_genomic.fna</td>\n",
       "      <td>LT598485.1</td>\n",
       "      <td>406885</td>\n",
       "      <td>408273</td>\n",
       "      <td>HRR25_61.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1389</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>yHAB155_kazachstania_spencerorum_160519.fas</td>\n",
       "      <td>scf7180000055587</td>\n",
       "      <td>23649</td>\n",
       "      <td>25022</td>\n",
       "      <td>HRR25_15.1</td>\n",
       "      <td>1</td>\n",
       "      <td>1374</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Genome             Contig q_start  \\\n",
       "0     GCA_030569995.1_ASM3056999v1_genomic.fna  JAJLUF010000032.1    8108   \n",
       "1                   tetrapisispora_phaffii.fas        NC_016524.1  591805   \n",
       "2                   tetrapisispora_phaffii.fas        NC_016529.1  523631   \n",
       "3            GCA_900074765.1_LAFE0_genomic.fna         LT598485.1  406885   \n",
       "4  yHAB155_kazachstania_spencerorum_160519.fas   scf7180000055587   23649   \n",
       "\n",
       "    q_end     Subject s_start s_end strand  \n",
       "0    9496  HRR25_60.1    1389     1     -1  \n",
       "1  593361  HRR25_35.1       1  1557      1  \n",
       "2  525016  HRR25_35.2    1386     1     -1  \n",
       "3  408273  HRR25_61.1       1  1389      1  \n",
       "4   25022  HRR25_15.1       1  1374      1  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions_of_interest = pd.DataFrame( columns=[\"Genome\",\"Contig\",\"q_start\",\"q_end\",\"Subject\",\"s_start\",\"s_end\",\"strand\"])\n",
    "\n",
    "for genome in blast_d:\n",
    "    for hit in blast_d[genome]:\n",
    "        el = blast_d[genome][hit]\n",
    "        identity = el[\"pident\"]\n",
    "\n",
    "        if identity == 100:\n",
    "            contig   =  el[\"contig\"]\n",
    "            str_on_q = el[\"str_q\"]\n",
    "            end_on_q = el[\"end_q\"]\n",
    "            subject  =  el[\"subject\"]\n",
    "            str_on_s = el[\"str_s\"]\n",
    "            end_on_s = el[\"end_s\"]\n",
    "            \n",
    "            if str_on_s < end_on_s:\n",
    "                strand = 1\n",
    "            elif str_on_s > end_on_s:\n",
    "                strand = -1\n",
    "        \n",
    "    \n",
    "        \n",
    "        \n",
    "            new_row  = {'Genome':genome.split(\".blast\")[0], \n",
    "                        'Contig':contig, \n",
    "                        'q_start':str_on_q, \n",
    "                        'q_end':end_on_q,\n",
    "                        'Subject':subject,\n",
    "                        's_start':str_on_s, \n",
    "                        's_end':end_on_s,\n",
    "                        'strand':strand\n",
    "                       }\n",
    "            regions_of_interest = regions_of_interest.append(new_row, ignore_index=True)\n",
    "            \n",
    "regions_of_interest.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "cc4cbb3e-a4cb-43be-adcf-c1a66e6f3645",
   "metadata": {},
   "outputs": [],
   "source": [
    "regions_of_interest.to_csv(\".blast_hits.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
