{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ab04e25f-95f9-4e89-b229-d918108c4294",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(readr)\n",
    "library(readxl)\n",
    "library(seqinr)\n",
    "# script shared from David Bradley"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "488d8663-d041-4d7b-967e-84f83b6497dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta <- read.fasta('HRR25.fasta',seqtype='AA')\n",
    "seq <- getSequence(fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c2a89619-aa8f-465c-8cb5-80d0d2fb96f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'S.cerevisiae_YPL204W_DupA'</li><li>'Z.rouxii_ZYRO0G08866g_Anc'</li><li>'T.delbrueckii_TDEL0E05230_Anc'</li><li>'N.castellii_NCAS0A01090_DupA'</li><li>'N.castellii_NCAS0D03820_DupB'</li><li>'V.polyspora_Kpol_1002.53_DupA'</li><li>'V.polyspora_Kpol_1033.51_DupB'</li><li>'T.phaffii_TPHA0E02830_DupA'</li><li>'T.phaffii_TPHA0J02370_DupB'</li><li>'K.naganishii_KNAG0F01980_DupB'</li><li>'K.naganishii_KNAG0D01130_DupA'</li><li>'K.africana_KAFR0C02290_DupB'</li><li>'K.africana_KAFR0K00320_DupA'</li><li>'N.dairenensis_NDAI0I01070_DupB'</li><li>'N.dairenensis_NDAI0F01640_DupA'</li><li>'T.blattae_TBLA0B05310_DupB'</li><li>'T.blattae_TBLA0A02040_DupA'</li><li>'S.bayanus'</li><li>'S.kudriavzevii_Skud_16.75_DupA'</li><li>'S.mikatae_Smik_6.409_DupA'</li><li>'C.glabrata_CAGL0H03553g_DupB'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'S.cerevisiae\\_YPL204W\\_DupA'\n",
       "\\item 'Z.rouxii\\_ZYRO0G08866g\\_Anc'\n",
       "\\item 'T.delbrueckii\\_TDEL0E05230\\_Anc'\n",
       "\\item 'N.castellii\\_NCAS0A01090\\_DupA'\n",
       "\\item 'N.castellii\\_NCAS0D03820\\_DupB'\n",
       "\\item 'V.polyspora\\_Kpol\\_1002.53\\_DupA'\n",
       "\\item 'V.polyspora\\_Kpol\\_1033.51\\_DupB'\n",
       "\\item 'T.phaffii\\_TPHA0E02830\\_DupA'\n",
       "\\item 'T.phaffii\\_TPHA0J02370\\_DupB'\n",
       "\\item 'K.naganishii\\_KNAG0F01980\\_DupB'\n",
       "\\item 'K.naganishii\\_KNAG0D01130\\_DupA'\n",
       "\\item 'K.africana\\_KAFR0C02290\\_DupB'\n",
       "\\item 'K.africana\\_KAFR0K00320\\_DupA'\n",
       "\\item 'N.dairenensis\\_NDAI0I01070\\_DupB'\n",
       "\\item 'N.dairenensis\\_NDAI0F01640\\_DupA'\n",
       "\\item 'T.blattae\\_TBLA0B05310\\_DupB'\n",
       "\\item 'T.blattae\\_TBLA0A02040\\_DupA'\n",
       "\\item 'S.bayanus'\n",
       "\\item 'S.kudriavzevii\\_Skud\\_16.75\\_DupA'\n",
       "\\item 'S.mikatae\\_Smik\\_6.409\\_DupA'\n",
       "\\item 'C.glabrata\\_CAGL0H03553g\\_DupB'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'S.cerevisiae_YPL204W_DupA'\n",
       "2. 'Z.rouxii_ZYRO0G08866g_Anc'\n",
       "3. 'T.delbrueckii_TDEL0E05230_Anc'\n",
       "4. 'N.castellii_NCAS0A01090_DupA'\n",
       "5. 'N.castellii_NCAS0D03820_DupB'\n",
       "6. 'V.polyspora_Kpol_1002.53_DupA'\n",
       "7. 'V.polyspora_Kpol_1033.51_DupB'\n",
       "8. 'T.phaffii_TPHA0E02830_DupA'\n",
       "9. 'T.phaffii_TPHA0J02370_DupB'\n",
       "10. 'K.naganishii_KNAG0F01980_DupB'\n",
       "11. 'K.naganishii_KNAG0D01130_DupA'\n",
       "12. 'K.africana_KAFR0C02290_DupB'\n",
       "13. 'K.africana_KAFR0K00320_DupA'\n",
       "14. 'N.dairenensis_NDAI0I01070_DupB'\n",
       "15. 'N.dairenensis_NDAI0F01640_DupA'\n",
       "16. 'T.blattae_TBLA0B05310_DupB'\n",
       "17. 'T.blattae_TBLA0A02040_DupA'\n",
       "18. 'S.bayanus'\n",
       "19. 'S.kudriavzevii_Skud_16.75_DupA'\n",
       "20. 'S.mikatae_Smik_6.409_DupA'\n",
       "21. 'C.glabrata_CAGL0H03553g_DupB'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"S.cerevisiae_YPL204W_DupA\"      \"Z.rouxii_ZYRO0G08866g_Anc\"     \n",
       " [3] \"T.delbrueckii_TDEL0E05230_Anc\"  \"N.castellii_NCAS0A01090_DupA\"  \n",
       " [5] \"N.castellii_NCAS0D03820_DupB\"   \"V.polyspora_Kpol_1002.53_DupA\" \n",
       " [7] \"V.polyspora_Kpol_1033.51_DupB\"  \"T.phaffii_TPHA0E02830_DupA\"    \n",
       " [9] \"T.phaffii_TPHA0J02370_DupB\"     \"K.naganishii_KNAG0F01980_DupB\" \n",
       "[11] \"K.naganishii_KNAG0D01130_DupA\"  \"K.africana_KAFR0C02290_DupB\"   \n",
       "[13] \"K.africana_KAFR0K00320_DupA\"    \"N.dairenensis_NDAI0I01070_DupB\"\n",
       "[15] \"N.dairenensis_NDAI0F01640_DupA\" \"T.blattae_TBLA0B05310_DupB\"    \n",
       "[17] \"T.blattae_TBLA0A02040_DupA\"     \"S.bayanus\"                     \n",
       "[19] \"S.kudriavzevii_Skud_16.75_DupA\" \"S.mikatae_Smik_6.409_DupA\"     \n",
       "[21] \"C.glabrata_CAGL0H03553g_DupB\"  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "names(fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b33bc82-432b-45ec-8981-f37a871df74e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pwm_files <- list.files(path='./pwm_dir/')\n",
    "\n",
    "# Read through each PWM file to generate a list of PWMs\n",
    "\n",
    "pwm_list <- NULL\n",
    "for (i in 1:length(pwm_files)) {\n",
    "  \n",
    "  path <- paste('./pwm_dir/',pwm_files[i],sep='')\n",
    "  \n",
    "  pwm <- read.table(path)\n",
    "  rownames(pwm) <- pwm[,1]\n",
    "  colnames(pwm) <- NULL\n",
    "  pwm <- pwm[,-1]\n",
    "  \n",
    "  pwm_list <- c(pwm_list, list(pwm))\n",
    "  \n",
    "}\n",
    "\n",
    "names(pwm_list) <- pwm_files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4f64c934-6801-456f-a09f-58e5bd72b090",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'P11710'</li><li>'P15891'</li><li>'P29366-1'</li><li>'P29366-2_PXXP'</li><li>'P32790-1_classI'</li><li>'P32790-2_classII'</li><li>'P32790-3'</li><li>'P32793'</li><li>'P36006'</li><li>'P38041'</li><li>'P38753'</li><li>'P38822-1'</li><li>'P38822-2'</li><li>'P39743_ClassI'</li><li>'P39743_ClassII'</li><li>'P39969'</li><li>'P40073'</li><li>'P43603'</li><li>'P47068_classIIcombined'</li><li>'P53281_classI'</li><li>'P53281_classII'</li><li>'P80667_classIIA'</li><li>'P80667_classIIB'</li><li>'Q04439'</li><li>'Q05080'</li><li>'Q06449_classI'</li><li>'Q06449_classII'</li><li>'Q07533_classI'</li><li>'Q07533_classII'</li><li>'Q12163_PXXP'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'P11710'\n",
       "\\item 'P15891'\n",
       "\\item 'P29366-1'\n",
       "\\item 'P29366-2\\_PXXP'\n",
       "\\item 'P32790-1\\_classI'\n",
       "\\item 'P32790-2\\_classII'\n",
       "\\item 'P32790-3'\n",
       "\\item 'P32793'\n",
       "\\item 'P36006'\n",
       "\\item 'P38041'\n",
       "\\item 'P38753'\n",
       "\\item 'P38822-1'\n",
       "\\item 'P38822-2'\n",
       "\\item 'P39743\\_ClassI'\n",
       "\\item 'P39743\\_ClassII'\n",
       "\\item 'P39969'\n",
       "\\item 'P40073'\n",
       "\\item 'P43603'\n",
       "\\item 'P47068\\_classIIcombined'\n",
       "\\item 'P53281\\_classI'\n",
       "\\item 'P53281\\_classII'\n",
       "\\item 'P80667\\_classIIA'\n",
       "\\item 'P80667\\_classIIB'\n",
       "\\item 'Q04439'\n",
       "\\item 'Q05080'\n",
       "\\item 'Q06449\\_classI'\n",
       "\\item 'Q06449\\_classII'\n",
       "\\item 'Q07533\\_classI'\n",
       "\\item 'Q07533\\_classII'\n",
       "\\item 'Q12163\\_PXXP'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'P11710'\n",
       "2. 'P15891'\n",
       "3. 'P29366-1'\n",
       "4. 'P29366-2_PXXP'\n",
       "5. 'P32790-1_classI'\n",
       "6. 'P32790-2_classII'\n",
       "7. 'P32790-3'\n",
       "8. 'P32793'\n",
       "9. 'P36006'\n",
       "10. 'P38041'\n",
       "11. 'P38753'\n",
       "12. 'P38822-1'\n",
       "13. 'P38822-2'\n",
       "14. 'P39743_ClassI'\n",
       "15. 'P39743_ClassII'\n",
       "16. 'P39969'\n",
       "17. 'P40073'\n",
       "18. 'P43603'\n",
       "19. 'P47068_classIIcombined'\n",
       "20. 'P53281_classI'\n",
       "21. 'P53281_classII'\n",
       "22. 'P80667_classIIA'\n",
       "23. 'P80667_classIIB'\n",
       "24. 'Q04439'\n",
       "25. 'Q05080'\n",
       "26. 'Q06449_classI'\n",
       "27. 'Q06449_classII'\n",
       "28. 'Q07533_classI'\n",
       "29. 'Q07533_classII'\n",
       "30. 'Q12163_PXXP'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"P11710\"                 \"P15891\"                 \"P29366-1\"              \n",
       " [4] \"P29366-2_PXXP\"          \"P32790-1_classI\"        \"P32790-2_classII\"      \n",
       " [7] \"P32790-3\"               \"P32793\"                 \"P36006\"                \n",
       "[10] \"P38041\"                 \"P38753\"                 \"P38822-1\"              \n",
       "[13] \"P38822-2\"               \"P39743_ClassI\"          \"P39743_ClassII\"        \n",
       "[16] \"P39969\"                 \"P40073\"                 \"P43603\"                \n",
       "[19] \"P47068_classIIcombined\" \"P53281_classI\"          \"P53281_classII\"        \n",
       "[22] \"P80667_classIIA\"        \"P80667_classIIB\"        \"Q04439\"                \n",
       "[25] \"Q05080\"                 \"Q06449_classI\"          \"Q06449_classII\"        \n",
       "[28] \"Q07533_classI\"          \"Q07533_classII\"         \"Q12163_PXXP\"           "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "names(pwm_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "85d81220-1959-46e5-b809-5c5d41be4694",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# Now, the PWM IDs need to be mapped to gene names\n",
    "\n",
    "pwm_ids <- rapply(strsplit(rapply(strsplit(names(pwm_list),split='_'), function(x) x[1]),split='-'), function(x) x[1])\n",
    "\n",
    "# Again, I will write out a file and then retrieve the gene names using the Uniprot ID mapper\n",
    "\n",
    "#write.table(pwm_ids,file='sh3_pwm_accessions_uniprot.txt',col.names=FALSE, row.names=FALSE, quote=FALSE)\n",
    "\n",
    "# Read in mapping table and process the strings to generate new names for the PWMs\n",
    "\n",
    "yeast_sh3_map <- read.table('./yeast_sh3_accession_to_GN.txt',header=TRUE,stringsAsFactors = FALSE)\n",
    "id_match <- yeast_sh3_map[match(pwm_ids,yeast_sh3_map[,1]),2]\n",
    "pwm_names <- paste(id_match,substr(names(pwm_list),7,rapply(strsplit(names(pwm_list),split=''), function(x) length(x))),sep='')\n",
    "\n",
    "names(pwm_list) <- pwm_names\n",
    "\n",
    "\n",
    "### Here is the code for the MSS function that I use to score potential SH3 ligands\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "e537fde4-899f-467e-bb63-d8acc9c749c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "### Here is the code for the MSS function that I use to score potential SH3 ligands\n",
    "\n",
    "\n",
    "mss_score <- function(psite,pwm) {\n",
    "  \n",
    "  psite <- unlist(strsplit(psite,split=''))\n",
    "  pwm <- as.matrix(pwm)\n",
    "    \n",
    "  match.ic = apply(pwm, 2, function(col) sum(col * logb(20 * col), na.rm=T))\n",
    "  \n",
    "  \n",
    "  c=0\n",
    "  min=0\n",
    "  max = 0\n",
    "  \n",
    "  # Ignore the central residue\n",
    "  \n",
    "  for (i in 1:ncol(pwm)) {\n",
    "    \n",
    "    aa <- psite[i]  \n",
    "    \n",
    "    # For missing amino acids, we score the '_' position\n",
    "    # the same as we would for a minimum frequency AA\n",
    "    \n",
    "    if (aa == '_') {\n",
    "      current <- min(pwm[,i])*match.ic[i]\n",
    "    } else {\n",
    "      current <- pwm[rownames(pwm) %in% psite[i],i]*match.ic[i]\n",
    "    }\n",
    "    \n",
    "    minimum <- min(pwm[,i])*match.ic[i]\n",
    "    maximum <- max(pwm[,i])*match.ic[i]\n",
    "    \n",
    "    c = c+current\n",
    "    min = min+minimum\n",
    "    max = max+maximum\n",
    "    \n",
    "  }\n",
    "  \n",
    "  mss = (c-min)/(max-min)\n",
    "  \n",
    "  return(unname(mss))\n",
    "  \n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "edef488a-244d-43fc-8f6c-8151c24e58bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Randomisation\n",
    "\n",
    "### Randomly sample from the proteome\n",
    "\n",
    "AAs <- c('A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','Y')\n",
    "\n",
    "sc_proteome <- read.fasta('yeast_proteome_uniprot.fasta', seqtype='AA')\n",
    "sc_aa_tab <- table(unlist(getSequence(sc_proteome)))\n",
    "sc_aa_tab <- sc_aa_tab[match(AAs,names(sc_aa_tab))]\n",
    "sc_aa_freq <- sc_aa_tab/sum(sc_aa_tab)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "b94cd8c3-88f3-48fc-9dc9-80a90933300a",
   "metadata": {},
   "outputs": [],
   "source": [
    "###### Randomisation\n",
    "\n",
    "target_seqs <- NULL\n",
    "\n",
    "# The whole proteome\n",
    "target_seqs <- seq\n",
    "\n",
    "max_mss_vec <- NULL\n",
    "\n",
    "for (j in 1:length(target_seqs)) {\n",
    "  \n",
    "  #print(j)\n",
    "  \n",
    "  target_seq <- target_seqs[[j]]\n",
    "  \n",
    "  # Now use a completely randomised sequences of the same length\n",
    "  \n",
    "  target_seq <- sample(AAs, prob = sc_aa_freq, size = length(target_seq), replace=TRUE)\n",
    "  \n",
    "  start_site_vec <- NULL\n",
    "  end_site_vec <- NULL\n",
    "  \n",
    "  # Generate the start and end sites\n",
    "  \n",
    "  for (k in 1:ncol(pwm)) {\n",
    "    \n",
    "    start_site <- paste(paste(rep('_',ncol(pwm)-k),collapse=''),paste(target_seq[1:k],collapse=''),sep='')\n",
    "    start_site_vec <- c(start_site_vec, start_site)\n",
    "    \n",
    "    end_site <- paste(paste(target_seq[(length(target_seq)-k+1):length(target_seq)],collapse=''),paste(rep('_',ncol(pwm)-k),collapse=''),sep='')\n",
    "    end_site_vec <- c(end_site_vec, end_site)  \n",
    "    \n",
    "  }\n",
    "  \n",
    "  # Now extract all possile binding sites from the sequence\n",
    "  \n",
    "  middle_site_vec <- NULL\n",
    "  \n",
    "  for (k in 2:(length(target_seq)-k)) {\n",
    "    \n",
    "    middle_site <- paste(target_seq[k:(k+ncol(pwm)-1)],collapse='')\n",
    "    middle_site_vec <- c(middle_site_vec, middle_site)\n",
    "    \n",
    "  }\n",
    "  \n",
    "  candidate_sites <- NULL\n",
    "  candidate_sites <- c(start_site_vec,middle_site_vec,rev(end_site_vec))\n",
    "  \n",
    "  # Iterate through all of the candidate sites and score with the MSS\n",
    "  \n",
    "  mss_vec <- NULL\n",
    "  \n",
    "  for (k in 1:length(candidate_sites)) {\n",
    "    \n",
    "    mss <- mss_score(candidate_sites[k],pwm)\n",
    "    mss_vec <- c(mss_vec,mss)\n",
    "    \n",
    "  }\n",
    "  \n",
    "  max_mss <- round(as.numeric(max(mss_vec)),3)\n",
    "  max_mss_vec <- c(max_mss_vec, max_mss)\n",
    "}\n",
    "\n",
    "# Use a lenient 95% threshold (usually about 0.7)\n",
    "\n",
    "thresh <- quantile(max_mss_vec,probs=seq(0,1,0.01))[96]\n",
    "\n",
    "rand_max_mss_vec <- max_mss_vec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "id": "8444b321-26b3-450e-b4a9-bc80635e805b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>95%:</strong> 0.64"
      ],
      "text/latex": [
       "\\textbf{95\\textbackslash{}\\%:} 0.64"
      ],
      "text/markdown": [
       "**95%:** 0.64"
      ],
      "text/plain": [
       " 95% \n",
       "0.64 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "thresh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7d0b9ebe-a467-41de-a893-885df00609c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# iterte through proteins\n",
    " # iterate through pwms\n",
    "AAs <- c('A','C','D','E','F','G','H','I','K','L','M','N','P','Q','R','S','T','V','W','Y')\n",
    "\n",
    "target_seqs <- NULL\n",
    "\n",
    "# The whole proteome\n",
    "target_seqs <- seq\n",
    "\n",
    "max_mss_vec <- NULL\n",
    "top_site_vec <- NULL\n",
    "pos_vec <- NULL\n",
    "name_vec <- NULL\n",
    "id_vec <- NULL\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6037e8dd-2dc5-434e-8044-55ff840a3c94",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in sample(AAs, prob = sc_aa_freq, size = length(target_seq), replace = TRUE):  オブジェクト 'sc_aa_freq' がありません \n",
     "output_type": "error",
     "traceback": [
      "Error in sample(AAs, prob = sc_aa_freq, size = length(target_seq), replace = TRUE):  オブジェクト 'sc_aa_freq' がありません \nTraceback:\n",
      "1. sample(AAs, prob = sc_aa_freq, size = length(target_seq), replace = TRUE)",
      "2. sample.int(length(x), size, replace, prob)"
     ]
    }
   ],
   "source": [
    "target_seqs <- NULL\n",
    "\n",
    "\n",
    "target_seqs <- seq\n",
    "\n",
    "max_mss_vec <- NULL\n",
    "top_site_vec <- NULL\n",
    "pos_vec <- NULL\n",
    "name_vec <- NULL\n",
    "id_vec <- NULL\n",
    "\n",
    "SH3_PWM_df <- data.frame(Gene_name=character(),Start_End=character(),PWM_name=character(),PWM_hits=character(),MSS=numeric()  )\n",
    "\n",
    "for (z in 1:length(pwm_list)) {    \n",
    "\n",
    "  pwm <- pwm_list[[z]]\n",
    "  id <- names(pwm_list)[z]\n",
    "  #print(id)\n",
    "    \n",
    "  ###### Proteome randomisation\n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "  target_seqs <- NULL\n",
    "  \n",
    "  # The whole proteome\n",
    "  target_seqs <- seq\n",
    "  \n",
    "  max_mss_vec <- NULL\n",
    "  \n",
    "  # Run it twice so that there wil be around 12,000\n",
    "  # random values in the null idstribution\n",
    "  \n",
    "  for (l in 1:2) {\n",
    "  \n",
    "  for (j in 1:length(target_seqs)) {\n",
    "    \n",
    "    #print(j)\n",
    "    \n",
    "    target_seq <- target_seqs[[j]]\n",
    "    \n",
    "    # Now use a completely randomised sequences of the same length\n",
    "    \n",
    "    target_seq <- sample(AAs, prob = sc_aa_freq, size = length(target_seq), replace=TRUE)\n",
    "    \n",
    "    start_site_vec <- NULL\n",
    "    end_site_vec <- NULL\n",
    "    \n",
    "    # Generate the start and end sites\n",
    "    \n",
    "    for (k in 1:ncol(pwm)) {\n",
    "      \n",
    "      start_site <- paste(paste(rep('_',ncol(pwm)-k),collapse=''),paste(target_seq[1:k],collapse=''),sep='')\n",
    "      start_site_vec <- c(start_site_vec, start_site)\n",
    "      \n",
    "      end_site <- paste(paste(target_seq[(length(target_seq)-k+1):length(target_seq)],collapse=''),paste(rep('_',ncol(pwm)-k),collapse=''),sep='')\n",
    "      end_site_vec <- c(end_site_vec, end_site)  \n",
    "    }\n",
    "    \n",
    "    # Now extract all possile binding sites from the sequence\n",
    "    middle_site_vec <- NULL\n",
    "    \n",
    "    for (k in 2:(length(target_seq)-k)) {\n",
    "      \n",
    "      middle_site <- paste(target_seq[k:(k+ncol(pwm)-1)],collapse='')\n",
    "      middle_site_vec <- c(middle_site_vec, middle_site)\n",
    "      \n",
    "    }\n",
    "    \n",
    "    candidate_sites <- NULL\n",
    "    candidate_sites <- c(start_site_vec,middle_site_vec,rev(end_site_vec))\n",
    "    \n",
    "    # Iterate through all of the candidate sites and score with the MSS\n",
    "    \n",
    "    mss_vec <- NULL\n",
    "    \n",
    "    for (k in 1:length(candidate_sites)) {\n",
    "      \n",
    "      mss <- mss_score(candidate_sites[k],pwm)\n",
    "      mss_vec <- c(mss_vec,mss)\n",
    "      \n",
    "    }\n",
    "    \n",
    "    max_mss <- round(as.numeric(max(mss_vec)),3)\n",
    "    max_mss_vec <- c(max_mss_vec, max_mss)\n",
    "  }\n",
    "  \n",
    "  }  \n",
    "    \n",
    "  # Use a lenient 95% threshold (usually about 0.7)\n",
    "  \n",
    "  thresh <- quantile(max_mss_vec,probs=seq(0,1,0.01))[96]\n",
    "  \n",
    "  rand_max_mss_vec <- max_mss_vec\n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "  ########### Real data\n",
    "  \n",
    "  ## Now, iterate across target sequnces\n",
    "  \n",
    "  \n",
    "  \n",
    "  for (j in 1:length(target_seqs)) {\n",
    "    max_mss_vec <- NULL\n",
    "    top_site_vec <- NULL\n",
    "    pos_vec <- NULL\n",
    "\n",
    "    name_vec = NULL\n",
    "    id_vec   = NULL\n",
    "    start_site_vec <- NULL\n",
    "    end_site_vec <- NULL\n",
    "  \n",
    "    target_seq <- target_seqs[[j]]\n",
    "    \n",
    "    name <- names(fasta)[j]\n",
    "    #print(name)\n",
    "    \n",
    "    # Generate the start and end sites\n",
    "    \n",
    "    for (k in 1:ncol(pwm)) {\n",
    "      \n",
    "      start_site <- paste(paste(rep('_',ncol(pwm)-k),collapse=''),paste(target_seq[1:k],collapse=''),sep='')\n",
    "      start_site_vec <- c(start_site_vec, start_site)\n",
    "      \n",
    "      end_site <- paste(paste(target_seq[(length(target_seq)-k+1):length(target_seq)],collapse=''),paste(rep('_',ncol(pwm)-k),collapse=''),sep='')\n",
    "      end_site_vec <- c(end_site_vec, end_site)  \n",
    "      \n",
    "    }\n",
    "    \n",
    "    # Now extract all possile binding sites from the sequence\n",
    "    \n",
    "    middle_site_vec <- NULL\n",
    "    \n",
    "    for (k in 2:(length(target_seq)-k)) {\n",
    "      \n",
    "      middle_site <- paste(target_seq[k:(k+ncol(pwm)-1)],collapse='')\n",
    "      middle_site_vec <- c(middle_site_vec, middle_site)\n",
    "      \n",
    "    }\n",
    "    \n",
    "    candidate_sites <- NULL\n",
    "    candidate_sites <- c(start_site_vec,middle_site_vec,rev(end_site_vec))\n",
    "    \n",
    "    # Iterate through all of the candidate sites and score with the MSS\n",
    "    \n",
    "    mss_vec <- NULL\n",
    "    \n",
    "    for (k in 1:length(candidate_sites)) {\n",
    "      \n",
    "      mss <- mss_score(candidate_sites[k],pwm)\n",
    "      mss_vec <- c(mss_vec,mss)\n",
    "      \n",
    "    }\n",
    "    \n",
    "    # We either report the maximum MSS or multiple MSS if there are several MSS above the threshold\n",
    "    \n",
    "    if (max(mss_vec) <= thresh) {\n",
    "    \n",
    "      pos_end <- which(mss_vec==max(mss_vec))\n",
    "      pos_start <- pos_end-ncol(pwm)+1\n",
    "      pos <- paste(pos_start,'-',pos_end,sep='')\n",
    "      pos_vec <- c(pos_vec, pos)\n",
    "      \n",
    "      top_site <- candidate_sites[which(mss_vec==max(mss_vec))]\n",
    "      top_site_vec <- c(top_site_vec,top_site)\n",
    "      \n",
    "      max_mss <- round(mss_vec[mss_vec==max(mss_vec)],3)\n",
    "      max_mss_vec <- c(max_mss_vec, max_mss)\n",
    "        \n",
    "\n",
    "      \n",
    "    } else {\n",
    "      \n",
    "      pos_end <- which(mss_vec>=thresh)\n",
    "      pos_start <- pos_end-ncol(pwm)+1\n",
    "      pos <- paste(pos_start,'-',pos_end,sep='')\n",
    "      pos_vec <- c(pos_vec, pos)\n",
    "      \n",
    "      top_site <- candidate_sites[which(mss_vec>=thresh)]\n",
    "      top_site_vec <- c(top_site_vec,top_site)\n",
    "      \n",
    "      max_mss <- mss_vec[which(mss_vec>=thresh)]\n",
    "      max_mss_vec <- c(max_mss_vec, max_mss)  \n",
    "\n",
    "      \n",
    "    }  \n",
    "\n",
    "    name_vec <- c(name_vec, rep(name,length(pos_vec)))\n",
    "    id_vec <- c(id_vec, rep(id,length(pos_vec)))\n",
    "\n",
    "      \n",
    "    if(length(pos_vec) != length(max_mss_vec)) {stop()}\n",
    "    if(length(pos_vec) != length(id_vec)) {print(pos_vec)\n",
    "                                           print(id_vec)\n",
    "                                           stop()}\n",
    "    if(length(pos_vec) != length(name_vec)) {\n",
    "        print(pos_vec)\n",
    "        print(name_vec)\n",
    "        stop()}\n",
    "    \n",
    "  \n",
    "  \n",
    "  max_mss_mat <- cbind(max_mss_vec,names(seq))\n",
    "\n",
    "  temp_df <- data.frame(name_vec,pos_vec,id_vec,top_site_vec,max_mss_vec)\n",
    "  colnames(temp_df) <- c('Gene_name','Start_End','PWM_name','PWM_hits','MSS')  \n",
    "\n",
    "  SH3_PWM_df = rbind(SH3_PWM_df,temp_df)\n",
    "}\n",
    "}  \n",
    "   \n",
    "  file_name <- 'SH3_PWM_scan_HRR25Orthologs_MSS.csv'\n",
    "  \n",
    "  write_csv(SH3_PWM_df, file_name, col_names = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7da5aa9d-125d-4aac-8323-9240f195026c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 0 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>Gene_name</th><th scope=col>Start_End</th><th scope=col>PWM_name</th><th scope=col>PWM_hits</th><th scope=col>MSS</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 0 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " Gene\\_name & Start\\_End & PWM\\_name & PWM\\_hits & MSS\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <dbl>\\\\\n",
       "\\hline\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 0 × 5\n",
       "\n",
       "| Gene_name &lt;chr&gt; | Start_End &lt;chr&gt; | PWM_name &lt;chr&gt; | PWM_hits &lt;chr&gt; | MSS &lt;dbl&gt; |\n",
       "|---|---|---|---|---|\n",
       "\n"
      ],
      "text/plain": [
       "     Gene_name Start_End PWM_name PWM_hits MSS"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "head(SH3_PWM_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "id": "0f0ab239-fa77-4887-a3db-bddf91b88c6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Done\"\n"
     ]
    }
   ],
   "source": [
    "print(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "899ef096-def0-45a9-90a9-8a3636f5f1ff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
